{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cb060a60-5584-4691-b73f-84986b69f610",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:utils’:\n",
      "\n",
      "    findMatches\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(scRNAseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1df7ad34-87c5-4cf4-9e84-993ff221b67d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Brain resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b48f056a-cce3-414a-bcc9-337ff7777339",
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1) Search: only \"brain\", human datasets\n",
    "# brain <- searchDatasets(defineTextQuery(\"brain\", partial = TRUE))\n",
    "\n",
    "## 2) Tidy summar\n",
    "# unlist_first <- function(x) paste0(unlist(x), collapse = \", \")\n",
    "# summ <- data.frame(\n",
    "#  name    = brain$name,\n",
    "#  version = brain$version,\n",
    "#  title   = brain$title,\n",
    "#  taxonomy_id = sapply(brain$taxonomy_id, unlist_first),  # 9606 = human; 10090 = mouse\n",
    "#  genome      = sapply(brain$genome,      unlist_first),  # GRCh* human; GRCm* mouse\n",
    "#  n_genes = brain$rows,\n",
    "#  n_cells = brain$columns,\n",
    "#  assays  = sapply(brain$assays,  unlist_first),\n",
    "#  coldata = sapply(brain$column_annotations, unlist_first),\n",
    "#  stringsAsFactors = FALSE\n",
    "# )\n",
    "# summ <- summ[order(summ$name, summ$version), ]\n",
    "# summ\n",
    "# If you want only human brain:\n",
    "# subset(summ, grepl(\"9606\", taxonomy_id) | grepl(\"^GRCh\", genome))\n",
    "\n",
    "## 3) Pretty “details per study”\n",
    "# for (i in seq_len(nrow(brain))) {\n",
    "#  src <- tryCatch(as.data.frame(brain$sources[[i]]), error = function(e) NULL)\n",
    "#  src_str <- if (!is.null(src) && nrow(src)) {\n",
    "#    paste(sprintf(\"%s:%s\", src$provider, src$id), collapse = \" | \")\n",
    "#  } else NA_character_\n",
    "\n",
    "#  cat(sprintf(\"\\n[%d] %s (version %s)\\n\", i, brain$name[i], brain$version[i]))\n",
    "#  cat(sprintf(\"Title:   %s\\n\", brain$title[i]))\n",
    "#  cat(sprintf(\"Genome:  %s   |   Taxonomy: %s\\n\",\n",
    "#              paste(unlist(brain$genome[[i]]), collapse=\", \"),\n",
    "#              paste(unlist(brain$taxonomy_id[[i]]), collapse=\", \")))\n",
    "#  cat(sprintf(\"Cells:   %s   |   Genes: %s\\n\", brain$columns[i], brain$rows[i]))\n",
    "#  cat(sprintf(\"Assays:  %s\\n\",\n",
    "#              paste(unlist(brain$assays[[i]]), collapse=\", \")))\n",
    "#  cat(sprintf(\"colData: %s\\n\",\n",
    "#              paste(unlist(brain$column_annotations[[i]]), collapse=\", \")))\n",
    "#  cat(sprintf(\"Sources: %s\\n\", src_str))\n",
    "# }\n",
    "\n",
    "## 4) (Optional)Save the table\n",
    "# write.csv(summ, \"scRNAseq_brain_datasets.csv\", row.names = FALSE)\n",
    "\n",
    "# library(scRNAseq)\n",
    "# name <- \"zhong-prefrontal-2018\"\n",
    "# ver  <- \"2023-12-22\"\n",
    "\n",
    "# File-backed by default; set realize.assays=TRUE to load into RAM (dgCMatrix)\n",
    "# sce <- fetchDataset(name, ver, realize.assays = TRUE)\n",
    "# sce\n",
    "\n",
    "# assayNames(sce)              # counts\n",
    "# dim(sce)                     # 24153 genes x 2394 cells (per your summary)\n",
    "# names(colData(sce))          # developmental_stage, gender, sample, cell_types, week, ...\n",
    "\n",
    "# Peek at labels\n",
    "# head(unique(colData(sce)$cell_types))\n",
    "# table(colData(sce)$developmental_stage, useNA = \"ifany\")\n",
    "\n",
    "# lab <- colData(sce)[[\"cell_types\"]]\n",
    "# unique(lab)\n",
    "# [1] \"Neurons\"           \"GABAergic neurons\" \"Microglia\"        \n",
    "# [4] \"Stem cells\"        NA                  \"Astrocytes\"       \n",
    "# [7] \"OPC\"       \n",
    "\n",
    "# SeuratData humancortexref.SeuratData\t\n",
    "\n",
    "# library(SeuratData)\n",
    "# AvailableData() \n",
    "# humancortexref.SeuratData\thumancortexref\t1.0.0\tAzimuth Reference: humancortex\thuman\tmotor cortex\t76533\tcells\n",
    "\n",
    "# CellXGene \n",
    "\n",
    "# Seattle AZ Alzheimer Disease Atlas (SEA-AD) : \n",
    "# https://cellxgene.cziscience.com/collections/1ca90a2d-2943-483d-b678-b809bf464c30\n",
    "# https://cellxgene.cziscience.com/collections/283d65eb-dd53-496d-adb7-7570c7caa443\n",
    "\n",
    "# Human Brain Atlas \n",
    "# https://cellxgene.cziscience.com/collections/283d65eb-dd53-496d-adb7-7570c7caa443"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d0b50a6-5f80-4242-b2ef-df3e69ca280a",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"SingleR\")\n",
    "library(SingleCellExperiment)\n",
    "\n",
    "# sce <- as.SingleCellExperiment(obj)\n",
    "# ref <- celldex::HumanPrimaryCellAtlasData()\n",
    "\n",
    "# pred_clust <- SingleR(test = sce, \n",
    "#                        ref = ref, \n",
    "#                        labels = ref$label.main,\n",
    "#                        clusters = sce$seurat_clusters)\n",
    "\n",
    "base_dir <- \"/mnt/nfs/CX000008_DS1/projects/btanasa/brain_refs_scRNAseq\"\n",
    "\n",
    "# Human Motor Cortex : \n",
    "# https://portal.brain-map.org/atlases-and-data/rnaseq/human-m1-10x\n",
    "# https://celltypes.brain-map.org/rnaseq/human_m1_10x?selectedVisualization=Heatmap&colorByFeature=Cell+Type&colorByFeatureValue=GAD1\n",
    "# https://portal.brain-map.org/atlases-and-data/rnaseq/human-mtg-10x_sea-ad\n",
    "# https://cellxgene.cziscience.com/collections/1ca90a2d-2943-483d-b678-b809bf464c30\n",
    "\n",
    "# Mouse annotations\n",
    "zeisel   <- readRDS(file.path(base_dir, \"ZeiselBrainData.rds\"))\n",
    "tasic    <- readRDS(file.path(base_dir, \"TasicBrainData.rds\"))\n",
    "romanov  <- readRDS(file.path(base_dir, \"RomanovBrainData.rds\"))\n",
    "\n",
    "# sanity checks\n",
    "stopifnot(inherits(zeisel, \"SingleCellExperiment\"))\n",
    "stopifnot(inherits(tasic, \"SingleCellExperiment\"))\n",
    "stopifnot(inherits(romanov, \"SingleCellExperiment\"))\n",
    "\n",
    "show_sce <- function(sce, name) {\n",
    "  cat(\"\\n================ \", name, \" ================\\n\", sep = \"\")\n",
    "  cat(\"assays: \", paste(assayNames(sce), collapse = \", \"), \"\\n\", sep = \"\")\n",
    "\n",
    "## 1) logcounts\n",
    "  if (\"logcounts\" %in% assayNames(sce)) {\n",
    "    logc <- assay(sce, \"logcounts\")\n",
    "    cat(\"logcounts  dim: \", paste(dim(logc), collapse = \" x \"),\n",
    "        \"  class: \", class(logc), \"\\n\", sep = \"\")\n",
    "    print(as.matrix(logc[1:min(5, nrow(logc)), 1:min(5, ncol(logc))]))\n",
    "  } else {\n",
    "    cat(\"No 'logcounts' assay found.\\n\")\n",
    "  }\n",
    "\n",
    "## 2) raw counts (if available)\n",
    "  raw_name <- if (\"counts\" %in% assayNames(sce)) \"counts\" else NA\n",
    "  if (is.na(raw_name)) {\n",
    "    cat(\"No raw 'counts' assay in this reference.\\n\")\n",
    "  } else {\n",
    "    cnt <- assay(sce, raw_name)\n",
    "    cat(raw_name, \" dim: \", paste(dim(cnt), collapse = \" x \"),\n",
    "        \"  class: \", class(cnt), \"\\n\", sep = \"\")\n",
    "    print(as.matrix(cnt[1:min(5, nrow(cnt)), 1:min(5, ncol(cnt))]))\n",
    "  }\n",
    "\n",
    "## 3) colData\n",
    "  cd <- colData(sce)\n",
    "  cat(\"colData columns: \", paste(colnames(cd), collapse = \", \"), \"\\n\", sep = \"\")\n",
    "  print(head(as.data.frame(cd), 10))\n",
    " }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aeca6a62-f268-4f89-ae32-680eec8882e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================ ZeiselBrainData ================\n",
      "assays: counts\n",
      "No 'logcounts' assay found.\n",
      "counts dim: 20006 x 3005  class: dgCMatrix\n",
      "         1772071015_C02 1772071017_G12 1772071017_A05 1772071014_B06\n",
      "Tspan12               0              0              0              3\n",
      "Tshz1                 3              1              0              2\n",
      "Fnbp1l                3              1              6              4\n",
      "Adamts15              0              0              0              0\n",
      "Cldn12                1              1              1              0\n",
      "         1772067065_H06\n",
      "Tspan12               0\n",
      "Tshz1                 2\n",
      "Fnbp1l                1\n",
      "Adamts15              0\n",
      "Cldn12                0\n",
      "colData columns: tissue, group #, total mRNA mol, well, sex, age, diameter, level1class, level2class\n",
      "                       tissue group.. total.mRNA.mol well sex age diameter\n",
      "1772071015_C02       sscortex       1          21580   11   1  21     0.00\n",
      "1772071017_G12       sscortex       1          21748   95  -1  20     9.56\n",
      "1772071017_A05       sscortex       1          31642   33  -1  20    11.10\n",
      "1772071014_B06       sscortex       1          32916   42   1  21    11.70\n",
      "1772067065_H06       sscortex       1          21531   48   1  25    11.00\n",
      "1772071017_E02       sscortex       1          24799   13  -1  20    11.90\n",
      "1772067065_B07       sscortex       1          31406   50   1  25    11.30\n",
      "1772067060_B09 ca1hippocampus       1          20389   66  -1  23    10.90\n",
      "1772071014_E04       sscortex       1          23022   29   1  21    12.90\n",
      "1772071015_D04       sscortex       1          24184   28   1  21    11.20\n",
      "                level1class level2class\n",
      "1772071015_C02 interneurons       Int10\n",
      "1772071017_G12 interneurons       Int10\n",
      "1772071017_A05 interneurons        Int6\n",
      "1772071014_B06 interneurons       Int10\n",
      "1772067065_H06 interneurons        Int9\n",
      "1772071017_E02 interneurons        Int9\n",
      "1772067065_B07 interneurons       Int10\n",
      "1772067060_B09 interneurons        Int9\n",
      "1772071014_E04 interneurons       Int10\n",
      "1772071015_D04 interneurons       Int10\n",
      "\n",
      "================ TasicBrainData ================\n",
      "assays: counts\n",
      "No 'logcounts' assay found.\n",
      "counts dim: 24058 x 1809  class: matrixarray\n",
      "              Calb2_tdTpositive_cell_1 Calb2_tdTpositive_cell_2\n",
      "0610005C13Rik                     0.00                     0.00\n",
      "0610007C21Rik                   992.00                  2287.02\n",
      "0610007L01Rik                     2.57                   177.00\n",
      "0610007N19Rik                     0.00                     0.00\n",
      "0610007P08Rik                     0.00                     0.00\n",
      "              Calb2_tdTpositive_cell_3 Calb2_tdTpositive_cell_4\n",
      "0610005C13Rik                     0.00                        0\n",
      "0610007C21Rik                   491.78                     1932\n",
      "0610007L01Rik                     0.00                        1\n",
      "0610007N19Rik                     0.00                        0\n",
      "0610007P08Rik                     0.00                        0\n",
      "              Calb2_tdTpositive_cell_5\n",
      "0610005C13Rik                        0\n",
      "0610007C21Rik                     1425\n",
      "0610007L01Rik                        2\n",
      "0610007N19Rik                        0\n",
      "0610007P08Rik                        0\n",
      "colData columns: mouse_line, cre_driver_1, cre_driver_2, cre_reporter, dissection, tdTomato, pass_qc_checks, broad_type, core_intermediate, primary_type, secondary_type, aibs_vignette_id\n",
      "                          mouse_line   cre_driver_1 cre_driver_2   cre_reporter\n",
      "Calb2_tdTpositive_cell_1       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_2       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_3       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_4       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_5       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_6       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_7       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_8       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_9       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_10      Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "                          dissection tdTomato pass_qc_checks\n",
      "Calb2_tdTpositive_cell_1         All positive              Y\n",
      "Calb2_tdTpositive_cell_2         All positive              Y\n",
      "Calb2_tdTpositive_cell_3         All positive              Y\n",
      "Calb2_tdTpositive_cell_4         All positive              Y\n",
      "Calb2_tdTpositive_cell_5         All positive              Y\n",
      "Calb2_tdTpositive_cell_6         All positive              Y\n",
      "Calb2_tdTpositive_cell_7         All positive              Y\n",
      "Calb2_tdTpositive_cell_8         All positive              Y\n",
      "Calb2_tdTpositive_cell_9         All positive              Y\n",
      "Calb2_tdTpositive_cell_10        All positive              Y\n",
      "                                    broad_type core_intermediate primary_type\n",
      "Calb2_tdTpositive_cell_1     GABA-ergic Neuron              core   Vip Mybpc1\n",
      "Calb2_tdTpositive_cell_2     GABA-ergic Neuron              core    Vip Parm1\n",
      "Calb2_tdTpositive_cell_3  Glutamatergic Neuron              core     L4 Ctxn3\n",
      "Calb2_tdTpositive_cell_4     GABA-ergic Neuron              core     Vip Chat\n",
      "Calb2_tdTpositive_cell_5     GABA-ergic Neuron              core    Vip Parm1\n",
      "Calb2_tdTpositive_cell_6  Glutamatergic Neuron              core   L2/3 Ptgs2\n",
      "Calb2_tdTpositive_cell_7  Glutamatergic Neuron              core       L2 Ngb\n",
      "Calb2_tdTpositive_cell_8  Glutamatergic Neuron              core   L2/3 Ptgs2\n",
      "Calb2_tdTpositive_cell_9  Glutamatergic Neuron      intermediate     L4 Ctxn3\n",
      "Calb2_tdTpositive_cell_10 Glutamatergic Neuron              core       L2 Ngb\n",
      "                          secondary_type aibs_vignette_id\n",
      "Calb2_tdTpositive_cell_1                           A200_V\n",
      "Calb2_tdTpositive_cell_2                           A201_V\n",
      "Calb2_tdTpositive_cell_3                           A202_V\n",
      "Calb2_tdTpositive_cell_4                           A203_V\n",
      "Calb2_tdTpositive_cell_5                           A204_V\n",
      "Calb2_tdTpositive_cell_6                           A205_V\n",
      "Calb2_tdTpositive_cell_7                           A206_V\n",
      "Calb2_tdTpositive_cell_8                           A207_V\n",
      "Calb2_tdTpositive_cell_9      L2/3 Ptgs2           A272_V\n",
      "Calb2_tdTpositive_cell_10                          A273_V\n",
      "\n",
      "================ RomanovBrainData ================\n",
      "assays: counts\n",
      "No 'logcounts' assay found.\n",
      "counts dim: 24341 x 2881  class: dgCMatrix\n",
      "       1772058147_F02 1772096158_E08 1772096144_A05 1772092004_A05\n",
      "Xkr4                0              0              0              0\n",
      "Rp1                 0              0              0              0\n",
      "Sox17               0              0              0              0\n",
      "Mrpl15              0              0              0              0\n",
      "Lypla1              0              0              0              0\n",
      "       1772092004_B06\n",
      "Xkr4                0\n",
      "Rp1                 0\n",
      "Sox17               0\n",
      "Mrpl15              3\n",
      "Lypla1              0\n",
      "colData columns: level1 class, level2 class (neurons only), level2 cluster number (neurons only), age (days postnatal), sex, cell diameter, acute stress, total molecules\n",
      "               level1.class level2.class..neurons.only.\n",
      "1772058147_F02       oligos                        <NA>\n",
      "1772096158_E08       oligos                        <NA>\n",
      "1772096144_A05       oligos                        <NA>\n",
      "1772092004_A05       oligos                        <NA>\n",
      "1772092004_B06       oligos                        <NA>\n",
      "1772096173_F02       oligos                        <NA>\n",
      "1772092004_B11       oligos                        <NA>\n",
      "1772078030_A08       oligos                        <NA>\n",
      "1772092004_B08       oligos                        <NA>\n",
      "1772092009_B06       oligos                        <NA>\n",
      "               level2.cluster.number..neurons.only. age..days.postnatal.\n",
      "1772058147_F02                                   NA                  NaN\n",
      "1772096158_E08                                   NA                   21\n",
      "1772096144_A05                                   NA                   27\n",
      "1772092004_A05                                   NA                   29\n",
      "1772092004_B06                                   NA                   29\n",
      "1772096173_F02                                   NA                   28\n",
      "1772092004_B11                                   NA                   29\n",
      "1772078030_A08                                   NA                   22\n",
      "1772092004_B08                                   NA                   29\n",
      "1772092009_B06                                   NA                   29\n",
      "                   sex cell.diameter acute.stress total.molecules\n",
      "1772058147_F02 unknown     10.248590        FALSE            2844\n",
      "1772096158_E08    male     10.076550        FALSE            3483\n",
      "1772096144_A05    male      9.255676         TRUE            7107\n",
      "1772092004_A05    male     11.486110         TRUE            6149\n",
      "1772092004_B06    male      7.848220         TRUE            5216\n",
      "1772096173_F02    male      6.341238         TRUE            2012\n",
      "1772092004_B11    male     11.521180         TRUE            8273\n",
      "1772078030_A08  female     10.069880        FALSE            5090\n",
      "1772092004_B08    male      8.521878         TRUE            5582\n",
      "1772092009_B06  female      8.662745        FALSE            3291\n"
     ]
    }
   ],
   "source": [
    "## run one by one\n",
    "show_sce(zeisel,   \"ZeiselBrainData\")\n",
    "show_sce(tasic,    \"TasicBrainData\")\n",
    "show_sce(romanov,  \"RomanovBrainData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "58207fdb-cb7a-4eda-b628-32b0ae499690",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'tissue'</li><li>'group #'</li><li>'total mRNA mol'</li><li>'well'</li><li>'sex'</li><li>'age'</li><li>'diameter'</li><li>'level1class'</li><li>'level2class'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'tissue'\n",
       "\\item 'group \\#'\n",
       "\\item 'total mRNA mol'\n",
       "\\item 'well'\n",
       "\\item 'sex'\n",
       "\\item 'age'\n",
       "\\item 'diameter'\n",
       "\\item 'level1class'\n",
       "\\item 'level2class'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'tissue'\n",
       "2. 'group #'\n",
       "3. 'total mRNA mol'\n",
       "4. 'well'\n",
       "5. 'sex'\n",
       "6. 'age'\n",
       "7. 'diameter'\n",
       "8. 'level1class'\n",
       "9. 'level2class'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"tissue\"         \"group #\"        \"total mRNA mol\" \"well\"          \n",
       "[5] \"sex\"            \"age\"            \"diameter\"       \"level1class\"   \n",
       "[9] \"level2class\"   "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================ ZeiselBrainData ================\n",
      "assays: counts\n",
      "No 'logcounts' assay found.\n",
      "counts dim: 20006 x 3005  class: dgCMatrix\n",
      "         1772071015_C02 1772071017_G12 1772071017_A05 1772071014_B06\n",
      "Tspan12               0              0              0              3\n",
      "Tshz1                 3              1              0              2\n",
      "Fnbp1l                3              1              6              4\n",
      "Adamts15              0              0              0              0\n",
      "Cldn12                1              1              1              0\n",
      "         1772067065_H06\n",
      "Tspan12               0\n",
      "Tshz1                 2\n",
      "Fnbp1l                1\n",
      "Adamts15              0\n",
      "Cldn12                0\n",
      "colData columns: tissue, group #, total mRNA mol, well, sex, age, diameter, level1class, level2class\n",
      "                       tissue group.. total.mRNA.mol well sex age diameter\n",
      "1772071015_C02       sscortex       1          21580   11   1  21     0.00\n",
      "1772071017_G12       sscortex       1          21748   95  -1  20     9.56\n",
      "1772071017_A05       sscortex       1          31642   33  -1  20    11.10\n",
      "1772071014_B06       sscortex       1          32916   42   1  21    11.70\n",
      "1772067065_H06       sscortex       1          21531   48   1  25    11.00\n",
      "1772071017_E02       sscortex       1          24799   13  -1  20    11.90\n",
      "1772067065_B07       sscortex       1          31406   50   1  25    11.30\n",
      "1772067060_B09 ca1hippocampus       1          20389   66  -1  23    10.90\n",
      "1772071014_E04       sscortex       1          23022   29   1  21    12.90\n",
      "1772071015_D04       sscortex       1          24184   28   1  21    11.20\n",
      "                level1class level2class\n",
      "1772071015_C02 interneurons       Int10\n",
      "1772071017_G12 interneurons       Int10\n",
      "1772071017_A05 interneurons        Int6\n",
      "1772071014_B06 interneurons       Int10\n",
      "1772067065_H06 interneurons        Int9\n",
      "1772071017_E02 interneurons        Int9\n",
      "1772067065_B07 interneurons       Int10\n",
      "1772067060_B09 interneurons        Int9\n",
      "1772071014_E04 interneurons       Int10\n",
      "1772071015_D04 interneurons       Int10\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'interneurons'</li><li>'pyramidal SS'</li><li>'pyramidal CA1'</li><li>'oligodendrocytes'</li><li>'microglia'</li><li>'endothelial-mural'</li><li>'astrocytes_ependymal'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'interneurons'\n",
       "\\item 'pyramidal SS'\n",
       "\\item 'pyramidal CA1'\n",
       "\\item 'oligodendrocytes'\n",
       "\\item 'microglia'\n",
       "\\item 'endothelial-mural'\n",
       "\\item 'astrocytes\\_ependymal'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'interneurons'\n",
       "2. 'pyramidal SS'\n",
       "3. 'pyramidal CA1'\n",
       "4. 'oligodendrocytes'\n",
       "5. 'microglia'\n",
       "6. 'endothelial-mural'\n",
       "7. 'astrocytes_ependymal'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"interneurons\"         \"pyramidal SS\"         \"pyramidal CA1\"       \n",
       "[4] \"oligodendrocytes\"     \"microglia\"            \"endothelial-mural\"   \n",
       "[7] \"astrocytes_ependymal\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## run one by one\n",
    "colnames(colData(zeisel))\n",
    "show_sce(zeisel,   \"ZeiselBrainData\")\n",
    "unique(colData(zeisel)[, 8])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c1207b40-fdb4-4b9b-a1f1-e42deeab353c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'mouse_line'</li><li>'cre_driver_1'</li><li>'cre_driver_2'</li><li>'cre_reporter'</li><li>'dissection'</li><li>'tdTomato'</li><li>'pass_qc_checks'</li><li>'broad_type'</li><li>'core_intermediate'</li><li>'primary_type'</li><li>'secondary_type'</li><li>'aibs_vignette_id'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'mouse\\_line'\n",
       "\\item 'cre\\_driver\\_1'\n",
       "\\item 'cre\\_driver\\_2'\n",
       "\\item 'cre\\_reporter'\n",
       "\\item 'dissection'\n",
       "\\item 'tdTomato'\n",
       "\\item 'pass\\_qc\\_checks'\n",
       "\\item 'broad\\_type'\n",
       "\\item 'core\\_intermediate'\n",
       "\\item 'primary\\_type'\n",
       "\\item 'secondary\\_type'\n",
       "\\item 'aibs\\_vignette\\_id'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'mouse_line'\n",
       "2. 'cre_driver_1'\n",
       "3. 'cre_driver_2'\n",
       "4. 'cre_reporter'\n",
       "5. 'dissection'\n",
       "6. 'tdTomato'\n",
       "7. 'pass_qc_checks'\n",
       "8. 'broad_type'\n",
       "9. 'core_intermediate'\n",
       "10. 'primary_type'\n",
       "11. 'secondary_type'\n",
       "12. 'aibs_vignette_id'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"mouse_line\"        \"cre_driver_1\"      \"cre_driver_2\"     \n",
       " [4] \"cre_reporter\"      \"dissection\"        \"tdTomato\"         \n",
       " [7] \"pass_qc_checks\"    \"broad_type\"        \"core_intermediate\"\n",
       "[10] \"primary_type\"      \"secondary_type\"    \"aibs_vignette_id\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================ TasicBrainData ================\n",
      "assays: counts\n",
      "No 'logcounts' assay found.\n",
      "counts dim: 24058 x 1809  class: matrixarray\n",
      "              Calb2_tdTpositive_cell_1 Calb2_tdTpositive_cell_2\n",
      "0610005C13Rik                     0.00                     0.00\n",
      "0610007C21Rik                   992.00                  2287.02\n",
      "0610007L01Rik                     2.57                   177.00\n",
      "0610007N19Rik                     0.00                     0.00\n",
      "0610007P08Rik                     0.00                     0.00\n",
      "              Calb2_tdTpositive_cell_3 Calb2_tdTpositive_cell_4\n",
      "0610005C13Rik                     0.00                        0\n",
      "0610007C21Rik                   491.78                     1932\n",
      "0610007L01Rik                     0.00                        1\n",
      "0610007N19Rik                     0.00                        0\n",
      "0610007P08Rik                     0.00                        0\n",
      "              Calb2_tdTpositive_cell_5\n",
      "0610005C13Rik                        0\n",
      "0610007C21Rik                     1425\n",
      "0610007L01Rik                        2\n",
      "0610007N19Rik                        0\n",
      "0610007P08Rik                        0\n",
      "colData columns: mouse_line, cre_driver_1, cre_driver_2, cre_reporter, dissection, tdTomato, pass_qc_checks, broad_type, core_intermediate, primary_type, secondary_type, aibs_vignette_id\n",
      "                          mouse_line   cre_driver_1 cre_driver_2   cre_reporter\n",
      "Calb2_tdTpositive_cell_1       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_2       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_3       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_4       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_5       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_6       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_7       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_8       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_9       Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "Calb2_tdTpositive_cell_10      Calb2 Calb2-IRES-Cre              RCL-tdT (Ai14)\n",
      "                          dissection tdTomato pass_qc_checks\n",
      "Calb2_tdTpositive_cell_1         All positive              Y\n",
      "Calb2_tdTpositive_cell_2         All positive              Y\n",
      "Calb2_tdTpositive_cell_3         All positive              Y\n",
      "Calb2_tdTpositive_cell_4         All positive              Y\n",
      "Calb2_tdTpositive_cell_5         All positive              Y\n",
      "Calb2_tdTpositive_cell_6         All positive              Y\n",
      "Calb2_tdTpositive_cell_7         All positive              Y\n",
      "Calb2_tdTpositive_cell_8         All positive              Y\n",
      "Calb2_tdTpositive_cell_9         All positive              Y\n",
      "Calb2_tdTpositive_cell_10        All positive              Y\n",
      "                                    broad_type core_intermediate primary_type\n",
      "Calb2_tdTpositive_cell_1     GABA-ergic Neuron              core   Vip Mybpc1\n",
      "Calb2_tdTpositive_cell_2     GABA-ergic Neuron              core    Vip Parm1\n",
      "Calb2_tdTpositive_cell_3  Glutamatergic Neuron              core     L4 Ctxn3\n",
      "Calb2_tdTpositive_cell_4     GABA-ergic Neuron              core     Vip Chat\n",
      "Calb2_tdTpositive_cell_5     GABA-ergic Neuron              core    Vip Parm1\n",
      "Calb2_tdTpositive_cell_6  Glutamatergic Neuron              core   L2/3 Ptgs2\n",
      "Calb2_tdTpositive_cell_7  Glutamatergic Neuron              core       L2 Ngb\n",
      "Calb2_tdTpositive_cell_8  Glutamatergic Neuron              core   L2/3 Ptgs2\n",
      "Calb2_tdTpositive_cell_9  Glutamatergic Neuron      intermediate     L4 Ctxn3\n",
      "Calb2_tdTpositive_cell_10 Glutamatergic Neuron              core       L2 Ngb\n",
      "                          secondary_type aibs_vignette_id\n",
      "Calb2_tdTpositive_cell_1                           A200_V\n",
      "Calb2_tdTpositive_cell_2                           A201_V\n",
      "Calb2_tdTpositive_cell_3                           A202_V\n",
      "Calb2_tdTpositive_cell_4                           A203_V\n",
      "Calb2_tdTpositive_cell_5                           A204_V\n",
      "Calb2_tdTpositive_cell_6                           A205_V\n",
      "Calb2_tdTpositive_cell_7                           A206_V\n",
      "Calb2_tdTpositive_cell_8                           A207_V\n",
      "Calb2_tdTpositive_cell_9      L2/3 Ptgs2           A272_V\n",
      "Calb2_tdTpositive_cell_10                          A273_V\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Calb2'</li><li>'Chat'</li><li>'Chrna2'</li><li>'Ctgf'</li><li>'Cux2'</li><li>'Gad2'</li><li>'Htr3a'</li><li>'Ndnf'</li><li>'Nkx2-1'</li><li>'Nos1'</li><li>'Nr5a1'</li><li>'Ntsr1'</li><li>'Pvalb'</li><li>'PvalbD-Slc32a1'</li><li>'PvalbF-Gad2'</li><li>'Rbp4'</li><li>'Rorb'</li><li>'Scnn1a-Tg2'</li><li>'Scnn1a-Tg3'</li><li>'Slc17a6'</li><li>'Sst'</li><li>'Tac2'</li><li>'Vip'</li><li>'CAV_LP_Ipsi'</li><li>'CAV_VISp_Contra'</li><li>'Trib2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Calb2'\n",
       "\\item 'Chat'\n",
       "\\item 'Chrna2'\n",
       "\\item 'Ctgf'\n",
       "\\item 'Cux2'\n",
       "\\item 'Gad2'\n",
       "\\item 'Htr3a'\n",
       "\\item 'Ndnf'\n",
       "\\item 'Nkx2-1'\n",
       "\\item 'Nos1'\n",
       "\\item 'Nr5a1'\n",
       "\\item 'Ntsr1'\n",
       "\\item 'Pvalb'\n",
       "\\item 'PvalbD-Slc32a1'\n",
       "\\item 'PvalbF-Gad2'\n",
       "\\item 'Rbp4'\n",
       "\\item 'Rorb'\n",
       "\\item 'Scnn1a-Tg2'\n",
       "\\item 'Scnn1a-Tg3'\n",
       "\\item 'Slc17a6'\n",
       "\\item 'Sst'\n",
       "\\item 'Tac2'\n",
       "\\item 'Vip'\n",
       "\\item 'CAV\\_LP\\_Ipsi'\n",
       "\\item 'CAV\\_VISp\\_Contra'\n",
       "\\item 'Trib2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Calb2'\n",
       "2. 'Chat'\n",
       "3. 'Chrna2'\n",
       "4. 'Ctgf'\n",
       "5. 'Cux2'\n",
       "6. 'Gad2'\n",
       "7. 'Htr3a'\n",
       "8. 'Ndnf'\n",
       "9. 'Nkx2-1'\n",
       "10. 'Nos1'\n",
       "11. 'Nr5a1'\n",
       "12. 'Ntsr1'\n",
       "13. 'Pvalb'\n",
       "14. 'PvalbD-Slc32a1'\n",
       "15. 'PvalbF-Gad2'\n",
       "16. 'Rbp4'\n",
       "17. 'Rorb'\n",
       "18. 'Scnn1a-Tg2'\n",
       "19. 'Scnn1a-Tg3'\n",
       "20. 'Slc17a6'\n",
       "21. 'Sst'\n",
       "22. 'Tac2'\n",
       "23. 'Vip'\n",
       "24. 'CAV_LP_Ipsi'\n",
       "25. 'CAV_VISp_Contra'\n",
       "26. 'Trib2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Calb2\"           \"Chat\"            \"Chrna2\"          \"Ctgf\"           \n",
       " [5] \"Cux2\"            \"Gad2\"            \"Htr3a\"           \"Ndnf\"           \n",
       " [9] \"Nkx2-1\"          \"Nos1\"            \"Nr5a1\"           \"Ntsr1\"          \n",
       "[13] \"Pvalb\"           \"PvalbD-Slc32a1\"  \"PvalbF-Gad2\"     \"Rbp4\"           \n",
       "[17] \"Rorb\"            \"Scnn1a-Tg2\"      \"Scnn1a-Tg3\"      \"Slc17a6\"        \n",
       "[21] \"Sst\"             \"Tac2\"            \"Vip\"             \"CAV_LP_Ipsi\"    \n",
       "[25] \"CAV_VISp_Contra\" \"Trib2\"          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## run one by one\n",
    "colnames(colData(tasic))\n",
    "show_sce(tasic,   \"TasicBrainData\")\n",
    "unique(colData(tasic)[, 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d19af83f-3729-4a44-9c53-7293aa0180d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'level1 class'</li><li>'level2 class (neurons only)'</li><li>'level2 cluster number (neurons only)'</li><li>'age (days postnatal)'</li><li>'sex'</li><li>'cell diameter'</li><li>'acute stress'</li><li>'total molecules'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'level1 class'\n",
       "\\item 'level2 class (neurons only)'\n",
       "\\item 'level2 cluster number (neurons only)'\n",
       "\\item 'age (days postnatal)'\n",
       "\\item 'sex'\n",
       "\\item 'cell diameter'\n",
       "\\item 'acute stress'\n",
       "\\item 'total molecules'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'level1 class'\n",
       "2. 'level2 class (neurons only)'\n",
       "3. 'level2 cluster number (neurons only)'\n",
       "4. 'age (days postnatal)'\n",
       "5. 'sex'\n",
       "6. 'cell diameter'\n",
       "7. 'acute stress'\n",
       "8. 'total molecules'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"level1 class\"                        \n",
       "[2] \"level2 class (neurons only)\"         \n",
       "[3] \"level2 cluster number (neurons only)\"\n",
       "[4] \"age (days postnatal)\"                \n",
       "[5] \"sex\"                                 \n",
       "[6] \"cell diameter\"                       \n",
       "[7] \"acute stress\"                        \n",
       "[8] \"total molecules\"                     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "================ RomanovBrainData ================\n",
      "assays: counts\n",
      "No 'logcounts' assay found.\n",
      "counts dim: 24341 x 2881  class: dgCMatrix\n",
      "       1772058147_F02 1772096158_E08 1772096144_A05 1772092004_A05\n",
      "Xkr4                0              0              0              0\n",
      "Rp1                 0              0              0              0\n",
      "Sox17               0              0              0              0\n",
      "Mrpl15              0              0              0              0\n",
      "Lypla1              0              0              0              0\n",
      "       1772092004_B06\n",
      "Xkr4                0\n",
      "Rp1                 0\n",
      "Sox17               0\n",
      "Mrpl15              3\n",
      "Lypla1              0\n",
      "colData columns: level1 class, level2 class (neurons only), level2 cluster number (neurons only), age (days postnatal), sex, cell diameter, acute stress, total molecules\n",
      "               level1.class level2.class..neurons.only.\n",
      "1772058147_F02       oligos                        <NA>\n",
      "1772096158_E08       oligos                        <NA>\n",
      "1772096144_A05       oligos                        <NA>\n",
      "1772092004_A05       oligos                        <NA>\n",
      "1772092004_B06       oligos                        <NA>\n",
      "1772096173_F02       oligos                        <NA>\n",
      "1772092004_B11       oligos                        <NA>\n",
      "1772078030_A08       oligos                        <NA>\n",
      "1772092004_B08       oligos                        <NA>\n",
      "1772092009_B06       oligos                        <NA>\n",
      "               level2.cluster.number..neurons.only. age..days.postnatal.\n",
      "1772058147_F02                                   NA                  NaN\n",
      "1772096158_E08                                   NA                   21\n",
      "1772096144_A05                                   NA                   27\n",
      "1772092004_A05                                   NA                   29\n",
      "1772092004_B06                                   NA                   29\n",
      "1772096173_F02                                   NA                   28\n",
      "1772092004_B11                                   NA                   29\n",
      "1772078030_A08                                   NA                   22\n",
      "1772092004_B08                                   NA                   29\n",
      "1772092009_B06                                   NA                   29\n",
      "                   sex cell.diameter acute.stress total.molecules\n",
      "1772058147_F02 unknown     10.248590        FALSE            2844\n",
      "1772096158_E08    male     10.076550        FALSE            3483\n",
      "1772096144_A05    male      9.255676         TRUE            7107\n",
      "1772092004_A05    male     11.486110         TRUE            6149\n",
      "1772092004_B06    male      7.848220         TRUE            5216\n",
      "1772096173_F02    male      6.341238         TRUE            2012\n",
      "1772092004_B11    male     11.521180         TRUE            8273\n",
      "1772078030_A08  female     10.069880        FALSE            5090\n",
      "1772092004_B08    male      8.521878         TRUE            5582\n",
      "1772092009_B06  female      8.662745        FALSE            3291\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Calb2'</li><li>'Chat'</li><li>'Chrna2'</li><li>'Ctgf'</li><li>'Cux2'</li><li>'Gad2'</li><li>'Htr3a'</li><li>'Ndnf'</li><li>'Nkx2-1'</li><li>'Nos1'</li><li>'Nr5a1'</li><li>'Ntsr1'</li><li>'Pvalb'</li><li>'PvalbD-Slc32a1'</li><li>'PvalbF-Gad2'</li><li>'Rbp4'</li><li>'Rorb'</li><li>'Scnn1a-Tg2'</li><li>'Scnn1a-Tg3'</li><li>'Slc17a6'</li><li>'Sst'</li><li>'Tac2'</li><li>'Vip'</li><li>'CAV_LP_Ipsi'</li><li>'CAV_VISp_Contra'</li><li>'Trib2'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'Calb2'\n",
       "\\item 'Chat'\n",
       "\\item 'Chrna2'\n",
       "\\item 'Ctgf'\n",
       "\\item 'Cux2'\n",
       "\\item 'Gad2'\n",
       "\\item 'Htr3a'\n",
       "\\item 'Ndnf'\n",
       "\\item 'Nkx2-1'\n",
       "\\item 'Nos1'\n",
       "\\item 'Nr5a1'\n",
       "\\item 'Ntsr1'\n",
       "\\item 'Pvalb'\n",
       "\\item 'PvalbD-Slc32a1'\n",
       "\\item 'PvalbF-Gad2'\n",
       "\\item 'Rbp4'\n",
       "\\item 'Rorb'\n",
       "\\item 'Scnn1a-Tg2'\n",
       "\\item 'Scnn1a-Tg3'\n",
       "\\item 'Slc17a6'\n",
       "\\item 'Sst'\n",
       "\\item 'Tac2'\n",
       "\\item 'Vip'\n",
       "\\item 'CAV\\_LP\\_Ipsi'\n",
       "\\item 'CAV\\_VISp\\_Contra'\n",
       "\\item 'Trib2'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'Calb2'\n",
       "2. 'Chat'\n",
       "3. 'Chrna2'\n",
       "4. 'Ctgf'\n",
       "5. 'Cux2'\n",
       "6. 'Gad2'\n",
       "7. 'Htr3a'\n",
       "8. 'Ndnf'\n",
       "9. 'Nkx2-1'\n",
       "10. 'Nos1'\n",
       "11. 'Nr5a1'\n",
       "12. 'Ntsr1'\n",
       "13. 'Pvalb'\n",
       "14. 'PvalbD-Slc32a1'\n",
       "15. 'PvalbF-Gad2'\n",
       "16. 'Rbp4'\n",
       "17. 'Rorb'\n",
       "18. 'Scnn1a-Tg2'\n",
       "19. 'Scnn1a-Tg3'\n",
       "20. 'Slc17a6'\n",
       "21. 'Sst'\n",
       "22. 'Tac2'\n",
       "23. 'Vip'\n",
       "24. 'CAV_LP_Ipsi'\n",
       "25. 'CAV_VISp_Contra'\n",
       "26. 'Trib2'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"Calb2\"           \"Chat\"            \"Chrna2\"          \"Ctgf\"           \n",
       " [5] \"Cux2\"            \"Gad2\"            \"Htr3a\"           \"Ndnf\"           \n",
       " [9] \"Nkx2-1\"          \"Nos1\"            \"Nr5a1\"           \"Ntsr1\"          \n",
       "[13] \"Pvalb\"           \"PvalbD-Slc32a1\"  \"PvalbF-Gad2\"     \"Rbp4\"           \n",
       "[17] \"Rorb\"            \"Scnn1a-Tg2\"      \"Scnn1a-Tg3\"      \"Slc17a6\"        \n",
       "[21] \"Sst\"             \"Tac2\"            \"Vip\"             \"CAV_LP_Ipsi\"    \n",
       "[25] \"CAV_VISp_Contra\" \"Trib2\"          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "## run one by one\n",
    "colnames(colData(romanov))\n",
    "show_sce(romanov,   \"RomanovBrainData\")\n",
    "unique(colData(tasic)[, 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1d10c0e6-9de5-4eb6-b2d2-bdde3c81d07b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "==level1.class==\n",
      "[1] \"oligos\"      \"astrocytes\"  \"ependymal\"   \"microglia\"   \"vsm\"        \n",
      "[6] \"endothelial\" \"neurons\"    \n",
      "\n",
      "==level2.class..neurons.only.==\n",
      " [1] NA                                 \"Adcyap1 1 (Tac1)\"                \n",
      " [3] \"Adcyap1 2\"                        \"Avp 1, high\"                     \n",
      " [5] \"Avp 2, high\"                      \"Avp 3, medium\"                   \n",
      " [7] \"Dopamine 1\"                       \"Dopamine 2 (low VMAT2)\"          \n",
      " [9] \"Dopamine 3\"                       \"Dopamine 4\"                      \n",
      "[11] \"GABA 10\"                          \"GABA 11 (Nts 1)\"                 \n",
      "[13] \"GABA 12 (Nts 2)\"                  \"GABA 13 (Galanin)\"               \n",
      "[15] \"GABA 14 (Npy,Agrp)\"               \"GABA 15 (Npy-medium)\"            \n",
      "[17] \"GABA 2 (Gucy1a3)\"                 \"GABA 3 (Crh+/-, Lhx6)\"           \n",
      "[19] \"GABA 4 (Crh+/-,Pgr15l)\"           \"GABA 5 (Calcr, Lhx1)\"            \n",
      "[21] \"GABA 6 (Otof, Lhx1)\"              \"GABA 7 (Pomc+/-)\"                \n",
      "[23] \"GABA 8\"                           \"GABA 9\"                          \n",
      "[25] \"GABA1\"                            \"Gad-low, Gnrh-/+\"                \n",
      "[27] \"Ghrh\"                             \"Hcrt\"                            \n",
      "[29] \"Hmit+/-\"                          \"Npvf\"                            \n",
      "[31] \"Oxytocin 1\"                       \"Oxytocin 2\"                      \n",
      "[33] \"Oxytocin 3\"                       \"Oxytocin 4\"                      \n",
      "[35] \"Pmch\"                             \"Qrfp\"                            \n",
      "[37] \"Sst 1, low\"                       \"Sst 2, high\"                     \n",
      "[39] \"Sst 3, medium\"                    \"Trh 1 (low)\"                     \n",
      "[41] \"Trh 2 (medium)\"                   \"Trh 3 (high)\"                    \n",
      "[43] \"Vglut2 1 (Penk)\"                  \"Vglut2 10 (Morn4,Prrc2a)\"        \n",
      "[45] \"Vglut2 11\"                        \"Vglut2 12 (Mgat4b)\"              \n",
      "[47] \"Vglut2 13(Ninl,Rfx5,Zfp346)\"      \"Vglut2 14 (Col9a2)\"              \n",
      "[49] \"Vglut2 15 (Hcn1,6430411K18Rik)\"   \"Vglut2 16 (Gm5595, Tnr)\"         \n",
      "[51] \"Vglut2 17 (A930013F10Rik,Pou2f2)\" \"Vglut2 18 (Zfp458,Ppp1r12b)\"     \n",
      "[53] \"Vglut2 2 (Crh+/-)\"                \"Vglut2 3 (Crh-+/-, low)\"         \n",
      "[55] \"Vglut2 4\"                         \"Vglut2 5 (Myt1,Lhx9)\"            \n",
      "[57] \"Vglut2 6 (Prmt8,Ugdh)\"            \"Vglut2 7 (Pgam,Snx12)\"           \n",
      "[59] \"Vglut2 8\"                         \"Vglut2 9 (Gpr149)\"               \n",
      "[61] \"circadian 1 (Vip,Grp+/-)\"         \"circadian 2 (Nms,VIP+/-)\"        \n",
      "[63] \"circadian 3 (Per2)\"               \"uc\"                              \n",
      "\n",
      "==sex==\n",
      "[1] \"unknown\" \"male\"    \"female\" \n",
      "\n",
      "==acute.stress==\n",
      "[1] FALSE  TRUE\n"
     ]
    }
   ],
   "source": [
    "# Show unique values for categorical columns in colData\n",
    "df <- as.data.frame(SummarizedExperiment::colData(romanov))\n",
    "cat_cols <- names(Filter(function(x) is.factor(x) || is.character(x) || is.logical(x), df))\n",
    "\n",
    "invisible(lapply(cat_cols, function(nm) {\n",
    "  cat(\"\\n==\", nm, \"==\\n\", sep = \"\")\n",
    "  print(unique(df[[nm]]))\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3abd0731-d47e-4510-9649-1bb436463dd8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "==mouse_line==\n",
      " [1] \"Calb2\"           \"Chat\"            \"Chrna2\"          \"Ctgf\"           \n",
      " [5] \"Cux2\"            \"Gad2\"            \"Htr3a\"           \"Ndnf\"           \n",
      " [9] \"Nkx2-1\"          \"Nos1\"            \"Nr5a1\"           \"Ntsr1\"          \n",
      "[13] \"Pvalb\"           \"PvalbD-Slc32a1\"  \"PvalbF-Gad2\"     \"Rbp4\"           \n",
      "[17] \"Rorb\"            \"Scnn1a-Tg2\"      \"Scnn1a-Tg3\"      \"Slc17a6\"        \n",
      "[21] \"Sst\"             \"Tac2\"            \"Vip\"             \"CAV_LP_Ipsi\"    \n",
      "[25] \"CAV_VISp_Contra\" \"Trib2\"          \n",
      "\n",
      "==cre_driver_1==\n",
      " [1] \"Calb2-IRES-Cre\"   \"Chat-IRES-Cre\"    \"Chrna2-Cre_OE25\"  \"Ctgf-2A-dgCre\"   \n",
      " [5] \"Cux2-CreERT2\"     \"Gad2-IRES-Cre\"    \"Htr3a-Cre_NO152\"  \"Ndnf-IRES2-dgCre\"\n",
      " [9] \"Nkx2-1-CreERT2\"   \"Nos1-CreERT2\"     \"Nr5a1-Cre\"        \"Ntsr1-Cre_GN220\" \n",
      "[13] \"Pvalb-IRES-Cre\"   \"Pvalb-2A-DreO\"    \"Pvalb-2A-FlpO\"    \"Rbp4-Cre_KL100\"  \n",
      "[17] \"Rorb-IRES2-Cre\"   \"Scnn1a-Tg2-Cre\"   \"Scnn1a-Tg3-Cre\"   \"Slc17a6-IRES-Cre\"\n",
      "[21] \"Sst-IRES-Cre\"     \"Tac2-IRES2-Cre\"   \"Vip-IRES-Cre\"     \"CAV2-Cre\"        \n",
      "[25] \"Trib2-2A-CreERT2\"\n",
      "\n",
      "==cre_driver_2==\n",
      "[1] \"\"                 \"Slc32a1-IRES-Cre\" \"Gad2-IRES-Cre\"    NA                \n",
      "\n",
      "==cre_reporter==\n",
      "[1] \"RCL-tdT (Ai14)\"    \"RCRL-tdT (Ai66)\"   \"RCFL-tdT (Ai65)\"  \n",
      "[4] \"RCL-tdT\"           \"Snap25-LSL-2A-GFP\"\n",
      "\n",
      "==dissection==\n",
      " [1] \"All\"     \"upper\"   \"lower\"   \"L6\"      \"L6b\"     \"L4\"      \"L1-L2/3\"\n",
      " [8] \"L2/3\"    \"L1\"      \"L6a\"     \"L5\"      \"Cortex\" \n",
      "\n",
      "==tdTomato==\n",
      "[1] \"positive\" \"negative\" NA        \n",
      "\n",
      "==pass_qc_checks==\n",
      "[1] \"Y\" \"N\"\n",
      "\n",
      "==broad_type==\n",
      "[1] \"GABA-ergic Neuron\"              \"Glutamatergic Neuron\"          \n",
      "[3] \"Endothelial Cell\"               \"Astrocyte\"                     \n",
      "[5] \"Unclassified\"                   \"Microglia\"                     \n",
      "[7] \"Oligodendrocyte\"                \"Oligodendrocyte Precursor Cell\"\n",
      "\n",
      "==core_intermediate==\n",
      "[1] \"core\"         \"intermediate\" \"unclassified\"\n",
      "\n",
      "==primary_type==\n",
      " [1] \"Vip Mybpc1\"    \"Vip Parm1\"     \"L4 Ctxn3\"      \"Vip Chat\"     \n",
      " [5] \"L2/3 Ptgs2\"    \"L2 Ngb\"        \"Pvalb Gpx3\"    \"Ndnf Cxcl14\"  \n",
      " [9] \"Vip Gpc3\"      \"Vip Sncg\"      \"L5a Batf3\"     \"Endo Myl9\"    \n",
      "[13] \"Ndnf Car4\"     \"L4 Scnn1a\"     \"Astro Gja1\"    \"Sst Myh8\"     \n",
      "[17] \"Sst Chodl\"     \"Sst Th\"        \"Sst Tacstd2\"   \"Sncg\"         \n",
      "[21] \"Sst Cdk6\"      \"L5a Pde1c\"     \"Pvalb Wt1\"     \"L6a Mgp\"      \n",
      "[25] \"Endo Tbc1d4\"   \"L6b Serpinb11\" \"L6b Rgs12\"     \"Unclassified\" \n",
      "[29] \"L4 Arf5\"       \"L5a Hsd11b1\"   \"Pvalb Tacr3\"   \"L5a Tcerg1l\"  \n",
      "[33] \"Sst Cbln4\"     \"Micro Ctss\"    \"Igtp\"          \"Pvalb Cpne5\"  \n",
      "[37] \"Pvalb Rspo2\"   \"Oligo 96_Rik\"  \"Smad3\"         \"Oligo Opalin\" \n",
      "[41] \"Pvalb Tpbg\"    \"L6a Sla\"       \"L6a Syt17\"     \"L6a Car12\"    \n",
      "[45] \"OPC Pdgfra\"    \"Pvalb Obox3\"   \"L5b Tph2\"      \"L5 Ucma\"      \n",
      "[49] \"L5 Chrna6\"     \"L5b Cdh13\"    \n",
      "\n",
      "==secondary_type==\n",
      " [1] \"\"             \"L2/3 Ptgs2\"   \"Vip Parm1\"    \"L6a Sla\"      \"OPC Pdgfra\"  \n",
      " [6] \"Ndnf Cxcl14\"  \"Vip Gpc3\"     \"Vip Mybpc1\"   \"L4 Ctxn3\"     \"Pvalb Gpx3\"  \n",
      "[11] \"Sst Cbln4\"    \"L5a Tcerg1l\"  \"L5a Pde1c\"    \"Sst Myh8\"     \"Sst Cdk6\"    \n",
      "[16] \"L5a Batf3\"    \"Pvalb Rspo2\"  \"Pvalb Obox3\"  NA             \"L4 Scnn1a\"   \n",
      "[21] \"L4 Arf5\"      \"L2 Ngb\"       \"Pvalb Tacr3\"  \"Cd34\"         \"Vip Chat\"    \n",
      "[26] \"L6b Rgs12\"    \"Igtp\"         \"L5a Hsd11b1\"  \"L6a Mgp\"      \"L6a Syt17\"   \n",
      "[31] \"Pvalb Wt1\"    \"L5b Tph2\"     \"L5b Cdh13\"    \"L5 Ucma\"      \"Oligo Opalin\"\n",
      "[36] \"Sst Tacstd2\"  \"L6a Car12\"   \n",
      "\n",
      "==aibs_vignette_id==\n",
      "   [1] \"A200_V\"     \"A201_V\"     \"A202_V\"     \"A203_V\"     \"A204_V\"    \n",
      "   [6] \"A205_V\"     \"A206_V\"     \"A207_V\"     \"A272_V\"     \"A273_V\"    \n",
      "  [11] \"A274_V\"     \"A275_V\"     \"A276_V\"     \"A277_V\"     \"A278_V\"    \n",
      "  [16] \"A279_V\"     \"A1602_V\"    \"A1603_V\"    \"A1604_V\"    \"A1605_V\"   \n",
      "  [21] \"A1606_V\"    \"A1607_V\"    \"A1608_V\"    \"A1609_V\"    \"A1610_V\"   \n",
      "  [26] \"A1611_V\"    \"A1612_V\"    \"A1613_V\"    \"A1614_V\"    \"A1615_V\"   \n",
      "  [31] \"A1616_V\"    \"A1617_V\"    \"A1618_VU\"   \"A1619_VU\"   \"A1620_VU\"  \n",
      "  [36] \"A1621_VU\"   \"A1622_VU\"   \"A1623_VU\"   \"A1624_VU\"   \"A1625_VU\"  \n",
      "  [41] \"A1626_VU\"   \"A1627_VU\"   \"A1628_VU\"   \"A1629_VU\"   \"A1630_VU\"  \n",
      "  [46] \"A1631_VU\"   \"A1632_VU\"   \"A1633_VU\"   \"A1634_VL\"   \"A1635_VL\"  \n",
      "  [51] \"A1636_VL\"   \"A1637_VL\"   \"A1638_VL\"   \"A1639_VL\"   \"A1640_VL\"  \n",
      "  [56] \"A1641_VL\"   \"A1642_VL\"   \"A1643_VL\"   \"A1644_VL\"   \"A1645_VL\"  \n",
      "  [61] \"A1646_VL\"   \"A1647_VL\"   \"A1648_VL\"   \"A1649_VL\"   \"Z1454_V\"   \n",
      "  [66] \"Z1455_V\"    \"Z1456_V\"    \"Z1457_V\"    \"Z1458_V\"    \"Z1459_V\"   \n",
      "  [71] \"K1780_V6\"   \"K1781_V6\"   \"K1782_V6\"   \"K1783_V6\"   \"K1784_V6\"  \n",
      "  [76] \"K1785_V6\"   \"K1786_V6\"   \"K1787_V6\"   \"K1788_V6\"   \"K1789_V6\"  \n",
      "  [81] \"K1790_V6\"   \"K1791_V6\"   \"K1792_V6\"   \"K654_V5\"    \"K655_V5\"   \n",
      "  [86] \"K656_V5\"    \"K657_V5\"    \"K658_V5\"    \"K659_V5\"    \"K660_V5\"   \n",
      "  [91] \"K661_V5\"    \"K662_V5\"    \"K663_V5\"    \"K664_V5\"    \"K665_V5\"   \n",
      "  [96] \"K666_V5\"    \"K667_V5\"    \"K668_V5\"    \"K669_V5\"    \"K670_V5\"   \n",
      " [101] \"K671_V5\"    \"K673_V5\"    \"K674_V5\"    \"K675_V5\"    \"K676_V5\"   \n",
      " [106] \"K677_V5\"    \"K678_V5\"    \"K679_V5\"    \"K680_V5\"    \"K681_V5\"   \n",
      " [111] \"K682_V5\"    \"K683_V5\"    \"K686_V5\"    \"K687_V5\"    \"K688_V5\"   \n",
      " [116] \"K690_V5\"    \"K691_V5\"    \"K692_V5\"    \"K693_V5\"    \"K694_V5\"   \n",
      " [121] \"K695_V5\"    \"K696_V5\"    \"K697_V5\"    \"K698_V5\"    \"K699_V5\"   \n",
      " [126] \"K700_V5\"    \"K701_V5\"    \"K1185_V\"    \"K1186_V\"    \"K1187_V\"   \n",
      " [131] \"K1188_V\"    \"K1189_V\"    \"K1190_V\"    \"K1354_V\"    \"K1355_V\"   \n",
      " [136] \"K1356_V\"    \"K1357_V\"    \"K1358_V\"    \"K1359_V\"    \"K1360_V\"   \n",
      " [141] \"K1361_V\"    \"K1362_V\"    \"K1363_V\"    \"K1364_V\"    \"K1365_V\"   \n",
      " [146] \"K1366_V\"    \"K1367_V\"    \"K1368_V\"    \"K1369_V\"    \"K1370_V\"   \n",
      " [151] \"K1371_V\"    \"K1372_V\"    \"K1373_V\"    \"K1374_V\"    \"Ct1986_V\"  \n",
      " [156] \"Ct1987_V\"   \"Ct1988_V\"   \"Ct1989_V\"   \"Ct1990_V\"   \"Ct1991_V\"  \n",
      " [161] \"Ct1992_V\"   \"Ct1993_V\"   \"Ct1994_V\"   \"Ct1995_V\"   \"Ct1650_V6b\"\n",
      " [166] \"Ct1651_V6b\" \"Ct1652_V6b\" \"Ct1653_V6b\" \"Ct1654_V6b\" \"Ct1655_V6b\"\n",
      " [171] \"Ct1656_V6b\" \"Ct1657_V6b\" \"Ct1658_V6b\" \"Ct1659_V6b\" \"Ct1660_V6b\"\n",
      " [176] \"Ct1661_V6b\" \"Ct1663_V6b\" NA           \"C188_V4\"    \"C189_V4\"   \n",
      " [181] \"C190_V4\"    \"C191_V4\"    \"C192_V4\"    \"C193_V4\"    \"C194_V4\"   \n",
      " [186] \"C195_V4\"    \"C478_V4\"    \"C479_V4\"    \"C480_V4\"    \"C481_V4\"   \n",
      " [191] \"C482_V4\"    \"C483_V4\"    \"C484_V4\"    \"C485_V4\"    \"C486_V4\"   \n",
      " [196] \"C487_V4\"    \"C488_V4\"    \"C489_V4\"    \"C490_V4\"    \"C491_V4\"   \n",
      " [201] \"C492_V4\"    \"C493_V4\"    \"C494_V4\"    \"C495_V4\"    \"C496_V4\"   \n",
      " [206] \"C497_V4\"    \"C498_V4\"    \"C499_V4\"    \"C500_V4\"    \"C501_V4\"   \n",
      " [211] \"C502_V4\"    \"C503_V4\"    \"C504_V4\"    \"C505_V4\"    \"C506_V4\"   \n",
      " [216] \"C507_V4\"    \"C508_V4\"    \"C509_V4\"    \"C510_V4\"    \"C511_V4\"   \n",
      " [221] \"C512_V4\"    \"C513_V4\"    \"C514_V4\"    \"C515_V4\"    \"C516_V4\"   \n",
      " [226] \"C517_V4\"    \"C958_V23\"   \"C959_V23\"   \"C960_V234\"  \"C961_V234\" \n",
      " [231] \"C962_V234\"  \"C963_V23\"   \"C964_V234\"  \"C965_V234\"  \"C966_V234\" \n",
      " [236] \"C967_V234\"  \"C968_V234\"  \"C969_V234\"  \"C970_V234\"  \"C971_V234\" \n",
      " [241] \"C972_V234\"  \"C973_V234\"  \"C974_V234\"  \"C975_V234\"  \"C976_V234\" \n",
      " [246] \"C977_V234\"  \"C978_V234\"  \"C979_V234\"  \"C980_V234\"  \"C981_V234\" \n",
      " [251] \"C1245_V\"    \"C1246_V\"    \"C1247_V\"    \"C1248_V\"    \"C1249_V\"   \n",
      " [256] \"C1250_V\"    \"C1251_V\"    \"C1252_V\"    \"C1254_V\"    \"C1255_V\"   \n",
      " [261] \"C1256_V\"    \"C1257_V\"    \"C1258_V\"    \"C1259_V\"    \"C1260_V\"   \n",
      " [266] \"C1261_V\"    \"C1262_V\"    \"C1263_V\"    \"C1264_V\"    \"C1265_V\"   \n",
      " [271] \"C1266_V\"    \"C1267_V\"    \"C1268_V\"    \"C1269_V\"    \"C1270_V\"   \n",
      " [276] \"C1271_V\"    \"C1664_V23\"  \"C1665_V23\"  \"C1666_V23\"  \"C1667_V23\" \n",
      " [281] \"C1668_V23\"  \"C1669_V23\"  \"C1670_V23\"  \"C1671_V23\"  \"C1672_V23\" \n",
      " [286] \"C1673_V23\"  \"C1674_V23\"  \"C1675_V23\"  \"C1676_V23\"  \"C1677_V23\" \n",
      " [291] \"C1678_V23\"  \"C1679_V23\"  \"C88_V23\"    \"C89_V23\"    \"C90_V23\"   \n",
      " [296] \"C91_V23\"    \"C92_V23\"    \"C93_V23\"    \"C94_V23\"    \"C95_V23\"   \n",
      " [301] \"C1996n_V23\" \"C1997n_V23\" \"C1998n_V23\" \"C1999n_V23\" \"C2000n_V23\"\n",
      " [306] \"C2001n_V23\" \"C2002n_V23\" \"C2003n_V23\" \"C2004n_V23\" \"C2005n_V23\"\n",
      " [311] \"C2006n_V23\" \"C2007n_V23\" \"C2008n_V23\" \"C2009n_V23\" \"C2010n_V23\"\n",
      " [316] \"D1216_V\"    \"D1217_V\"    \"D1218_V\"    \"D1219_V\"    \"D1220_V\"   \n",
      " [321] \"D1221_V\"    \"D1222_V\"    \"D1223_V\"    \"D1224_V\"    \"D1225_V\"   \n",
      " [326] \"D1226_V\"    \"D1227_V\"    \"D1228_V\"    \"D1229_V\"    \"D1230_V\"   \n",
      " [331] \"D1231_V\"    \"D1232_V\"    \"D1233_V\"    \"D1234_V\"    \"D1235_V\"   \n",
      " [336] \"D1236_V\"    \"D1237_V\"    \"D1238_V\"    \"D1239_V\"    \"D1240_V\"   \n",
      " [341] \"D1241_V\"    \"D1242_V\"    \"D1243_V\"    \"D1244_V\"    \"D1306_V\"   \n",
      " [346] \"D1307_V\"    \"D1308_V\"    \"D1311_V\"    \"D1312_V\"    \"D1313_V\"   \n",
      " [351] \"D1314_V\"    \"D1315_V\"    \"D1316_V\"    \"D1317_V\"    \"D1318_V\"   \n",
      " [356] \"D1319_V\"    \"D1320_V\"    \"D1321_V\"    \"D1326_V\"    \"D1327_V\"   \n",
      " [361] \"D1328_V\"    \"D1329_V\"    \"D1330_V\"    \"D1331_V\"    \"D1332_V\"   \n",
      " [366] \"D1334_V\"    \"D1335_V\"    \"D1336_V\"    \"D1338_V\"    \"D1339_V\"   \n",
      " [371] \"D1340_V\"    \"D1341_V\"    \"D1342_V\"    \"D1343_V\"    \"D1344_V\"   \n",
      " [376] \"D1345_V\"    \"D1346_V\"    \"D1347_V\"    \"D1348_V\"    \"D1349_V\"   \n",
      " [381] \"D1350_V\"    \"D1351_V\"    \"D1352_V\"    \"D1353_V\"    \"D1378_V\"   \n",
      " [386] \"D1379_V\"    \"D1380_V\"    \"D1381_V\"    \"D1388_V\"    \"D1389_V\"   \n",
      " [391] \"D1390_V\"    \"D1391_V\"    \"H1006_V\"    \"H1007_V\"    \"H1008_V\"   \n",
      " [396] \"H1009_V\"    \"H1010_V\"    \"H1011_V\"    \"H1012_V\"    \"H1013_V\"   \n",
      " [401] \"H1014_V\"    \"H1015_V\"    \"H1016_V\"    \"H1017_V\"    \"H1018_V\"   \n",
      " [406] \"H1019_V\"    \"H1020_V\"    \"H1021_V\"    \"H1022_V\"    \"H1023_V\"   \n",
      " [411] \"H1024_V\"    \"H1025_V\"    \"H1026_V\"    \"H1027_V\"    \"H1028_V\"   \n",
      " [416] \"H1029_V\"    \"H1030_V\"    \"H1031_V\"    \"H1032_V\"    \"H1033_V\"   \n",
      " [421] \"H1034_V\"    \"H1035_V\"    \"H1036_V\"    \"H1037_V\"    \"H1038_V\"   \n",
      " [426] \"H1039_V\"    \"H1040_V\"    \"H1041_V\"    \"H1042_V\"    \"H1043_V\"   \n",
      " [431] \"H1044_V\"    \"H1045_V\"    \"H1046_V\"    \"H1047_V\"    \"H1048_V\"   \n",
      " [436] \"H1049_V\"    \"H1050_V\"    \"H1051_V\"    \"H1052_V\"    \"H1053_V\"   \n",
      " [441] \"H1099_VU\"   \"H1100_VU\"   \"H1101_VU\"   \"H1102_VU\"   \"H1103_VU\"  \n",
      " [446] \"H1104_VU\"   \"H1105_VU\"   \"H1106_VU\"   \"H1107_VU\"   \"H1108_VU\"  \n",
      " [451] \"H1109_VU\"   \"H1110_VU\"   \"H1111_VU\"   \"H1113_VU\"   \"H1114_VU\"  \n",
      " [456] \"H1115_VU\"   \"H1116_VU\"   \"H1117_VU\"   \"H1118_VU\"   \"H1119_VU\"  \n",
      " [461] \"H1120_VU\"   \"H1121_VU\"   \"H1122_VU\"   \"H1123_VL\"   \"H1124_VL\"  \n",
      " [466] \"H1125_VL\"   \"H1126_VL\"   \"H1127_VL\"   \"H1128_VL\"   \"H1129_VL\"  \n",
      " [471] \"H1130_VL\"   \"H1131_VL\"   \"H1132_VL\"   \"H1133_VL\"   \"H1134_VL\"  \n",
      " [476] \"H1135_VL\"   \"H1136_VL\"   \"H1137_VL\"   \"H1138_VL\"   \"H1139_VL\"  \n",
      " [481] \"H1140_VL\"   \"H1141_VL\"   \"H1142_VL\"   \"H1143_VL\"   \"H1144_VL\"  \n",
      " [486] \"H1145_VL\"   \"H1146_VL\"   \"H1392_V6b\"  \"H1393_V6b\"  \"H1394_V6b\" \n",
      " [491] \"H1395_V6b\"  \"H1396_V6b\"  \"H1397_V6b\"  \"H1398_V6b\"  \"H1399_V6b\" \n",
      " [496] \"H1400_V6b\"  \"H1401_V6b\"  \"H1402_V6b\"  \"H1403_V6b\"  \"H1407_V6b\" \n",
      " [501] \"H1408_V6b\"  \"H1409_V6b\"  \"H1410_V6b\"  \"H1411_V6b\"  \"H1412_V6b\" \n",
      " [506] \"H1413_V6b\"  \"H1414_V6b\"  \"H1415_V6b\"  \"H1416_V6b\"  \"H1417_V6b\" \n",
      " [511] \"H1418_V6b\"  \"H1419_V6b\"  \"H1420_V6b\"  \"H1421_V6b\"  \"H1422_V6b\" \n",
      " [516] \"Nd1970_V\"   \"Nd1971_V\"   \"Nd1972_V\"   \"Nd1977_V\"   \"Nd1978_V\"  \n",
      " [521] \"Nd1979_V\"   \"Nd1980_V\"   \"Nd1981_V\"   \"Nd1982_V\"   \"Nd1983_V\"  \n",
      " [526] \"Nd1984_V\"   \"Nd1985_V\"   \"Nd1680_V1\"  \"Nd1681_V1\"  \"Nd1682_V1\" \n",
      " [531] \"Nd1683_V1\"  \"Nd1684_V1\"  \"Nd1685_V1\"  \"Nd1686_V1\"  \"Nd1958n_V1\"\n",
      " [536] \"Nd1959n_V1\" \"Nd1960n_V1\" \"Nd1961n_V1\" \"Nd1962n_V1\" \"Nd1963n_V1\"\n",
      " [541] \"Nd1964n_V1\" \"Nd1965n_V1\" \"Nd1966n_V1\" \"Nd1967n_V1\" \"Nd1968n_V1\"\n",
      " [546] \"Nd1969n_V1\" \"W1793_VU\"   \"W1794_VU\"   \"W1795_VU\"   \"W1796_VU\"  \n",
      " [551] \"W1797_VU\"   \"W1798_VU\"   \"W1799_VU\"   \"W1800_VU\"   \"W1801_VU\"  \n",
      " [556] \"W1802_VU\"   \"W1803_VU\"   \"W1804_VU\"   \"W1805_VU\"   \"W1806_VU\"  \n",
      " [561] \"W1807_VU\"   \"W1808_VU\"   \"W1809_VU\"   \"W1810_VU\"   \"W1811_VU\"  \n",
      " [566] \"W1812_VU\"   \"W1813_VL\"   \"W1814_VL\"   \"U180_V\"     \"U181_V\"    \n",
      " [571] \"U182_V\"     \"U183_V\"     \"U184_V\"     \"U185_V\"     \"U186_V\"    \n",
      " [576] \"U187_V\"     \"U248_V\"     \"U249_V\"     \"U250_V\"     \"U251_V\"    \n",
      " [581] \"U252_V\"     \"U253_V\"     \"U254_V\"     \"U255_V\"     \"U726_V\"    \n",
      " [586] \"U727_V\"     \"U728_V\"     \"U729_V\"     \"U730_V\"     \"U731_V\"    \n",
      " [591] \"U1404_V\"    \"U1405_V\"    \"U1406_V\"    \"U1423_V\"    \"U1424_V\"   \n",
      " [596] \"U1426_V\"    \"U1427_V\"    \"U1428_V\"    \"U1460_V\"    \"U1461_V\"   \n",
      " [601] \"U1462_V\"    \"U1463_V\"    \"U1464_V\"    \"U1465_V\"    \"U1466_V\"   \n",
      " [606] \"U1467_V\"    \"U1468_V\"    \"U1469_V\"    \"U1470_V\"    \"U1471_V\"   \n",
      " [611] \"U1472_V\"    \"U1473_V\"    \"U1474_V\"    \"U1475_V\"    \"U1476_V\"   \n",
      " [616] \"U1477_V\"    \"U1479_V\"    \"U1480_V\"    \"U1481_V\"    \"U1482_V\"   \n",
      " [621] \"U1483_V\"    \"U1484_V\"    \"U1485_V\"    \"U1486_V\"    \"U1487_V\"   \n",
      " [626] \"U1488_V\"    \"U1489_V\"    \"U1490_V\"    \"U1491_V\"    \"U1497_V\"   \n",
      " [631] \"U1498_V\"    \"U1499_V\"    \"U1500_V\"    \"U1501_V\"    \"U1502_V\"   \n",
      " [636] \"U1503_V\"    \"U1504_V\"    \"U1505_V\"    \"U1506_V\"    \"U1507_V\"   \n",
      " [641] \"N232_V4\"    \"N233_V4\"    \"N234_V4\"    \"N235_V4\"    \"N236_V4\"   \n",
      " [646] \"N237_V4\"    \"N238_V4\"    \"N239_V4\"    \"N732_V4\"    \"N733_V4\"   \n",
      " [651] \"N939_V4\"    \"N940_V4\"    \"N941_V4\"    \"N942_V4\"    \"N943_V4\"   \n",
      " [656] \"N944_V4\"    \"N945_V4\"    \"N946_V4\"    \"N1548_V\"    \"N1549_V\"   \n",
      " [661] \"N1551_V\"    \"N1553_V\"    \"N1586_V\"    \"N1589_V\"    \"N1590_V\"   \n",
      " [666] \"N1591_V\"    \"N1592_V\"    \"N1593_V\"    \"N1711_V\"    \"N1712_V\"   \n",
      " [671] \"N1713_V\"    \"N1714_V\"    \"N1715_V\"    \"N1716_V\"    \"N1717_V\"   \n",
      " [676] \"N1718_V\"    \"N152_V4\"    \"N154_V4\"    \"N155_V4\"    \"N156_V4\"   \n",
      " [681] \"N157_V4\"    \"N158_V4\"    \"N159_V4\"    \"N347_V4\"    \"N348_V4\"   \n",
      " [686] \"N349_V4\"    \"N350_V4\"    \"N351_V4\"    \"T208_V6\"    \"T209_V6\"   \n",
      " [691] \"T210_V6\"    \"T211_V6\"    \"T212_V6\"    \"T213_V6\"    \"T214_V6\"   \n",
      " [696] \"T215_V6\"    \"T2127_V6a\"  \"T2128_V6a\"  \"T2129_V6a\"  \"T2130_V6a\" \n",
      " [701] \"T2131_V6a\"  \"T2132_V6a\"  \"T2133_V6a\"  \"T2134_V6a\"  \"T2135_V6a\" \n",
      " [706] \"T2136_V6a\"  \"T2143_V6a\"  \"T2144_V6a\"  \"T2145_V6a\"  \"T2146_V6a\" \n",
      " [711] \"T2147_V6a\"  \"T2148_V6a\"  \"T2149_V6a\"  \"T2150_V6a\"  \"T2151_V6a\" \n",
      " [716] \"T2152_V6a\"  \"T256_V6\"    \"T257_V6\"    \"T258_V6\"    \"T259_V6\"   \n",
      " [721] \"T260_V6\"    \"T261_V6\"    \"T262_V6\"    \"T263_V6\"    \"T446_V6a\"  \n",
      " [726] \"T447_V6a\"   \"T448_V6a\"   \"T449_V6a\"   \"T450_V6a\"   \"T451_V6a\"  \n",
      " [731] \"T452_V6a\"   \"T453_V6a\"   \"T454_V6a\"   \"T455_V6a\"   \"T456_V6a\"  \n",
      " [736] \"T457_V6a\"   \"T458_V6a\"   \"T459_V6a\"   \"T460_V6a\"   \"T461_V6a\"  \n",
      " [741] \"T702_V6\"    \"T703_V6\"    \"T704_V6\"    \"T707_V6\"    \"T708_V6\"   \n",
      " [746] \"T709_V6\"    \"T710_V6\"    \"T711_V6\"    \"T712_V6\"    \"T713_V6\"   \n",
      " [751] \"T714_V6\"    \"T715_V6\"    \"T716_V6\"    \"T717_V6\"    \"T718_V6\"   \n",
      " [756] \"T719_V6\"    \"T720_V6\"    \"T721_V6\"    \"T722_V6\"    \"T723_V6\"   \n",
      " [761] \"T724_V6\"    \"T725_V6\"    \"T128_V6\"    \"T129_V6\"    \"T130_V6\"   \n",
      " [766] \"T1816n_V6a\" \"T1817n_V6a\" \"T1818n_V6a\" \"T1819n_V6a\" \"T1820n_V6a\"\n",
      " [771] \"T1821n_V6a\" \"T1822n_V6a\" \"T1823n_V6a\" \"T1824n_V6a\" \"T1825n_V6a\"\n",
      " [776] \"T1826n_V6a\" \"T1827n_V6a\" \"T1828n_V6a\" \"T1829n_V6a\" \"T1830n_V6a\"\n",
      " [781] \"T1831n_V6a\" \"T1832n_V6a\" \"T1833n_V6a\" \"T1834n_V6a\" \"T1835n_V6a\"\n",
      " [786] \"T1836n_V6a\" \"T1837n_V6a\" \"T1838n_V6a\" \"T1839n_V6a\" \"T1840n_V6a\"\n",
      " [791] \"T1841n_V6a\" \"T1842n_V6a\" \"T1844n_V6a\" \"T1845n_V6a\" \"T1846n_V6a\"\n",
      " [796] \"T1847n_V6a\" \"T1848n_V6a\" \"T1849n_V6a\" \"T1850n_V6a\" \"T1852n_V6a\"\n",
      " [801] \"T1171n_V6\"  \"T1172n_V6\"  \"T1173n_V6\"  \"T1174n_V6\"  \"T1175n_V6\" \n",
      " [806] \"T1176n_V6\"  \"T1177n_V6\"  \"T1178n_V6\"  \"T1180n_V6\"  \"T1181n_V6\" \n",
      " [811] \"T1182n_V6\"  \"T1183n_V6\"  \"T1184n_V6\"  \"T1322n_V6\"  \"T1323n_V6\" \n",
      " [816] \"T1324n_V6\"  \"T1325n_V6\"  \"T1375n_V6\"  \"T1376n_V6\"  \"T1377n_V6\" \n",
      " [821] \"Y1900_VU\"   \"Y1901_VU\"   \"Y1902_VU\"   \"Y1903_VU\"   \"Y1904_VU\"  \n",
      " [826] \"Y1905_VU\"   \"Y1906_VU\"   \"Y1907_VU\"   \"Y1908_VU\"   \"Y1909_VU\"  \n",
      " [831] \"Y1910_VU\"   \"Y1911_VU\"   \"Y1912_VU\"   \"Y1913_VU\"   \"Y1914_VU\"  \n",
      " [836] \"Y1915_VU\"   \"Y1916_VU\"   \"Y1917_VU\"   \"Y1918_VU\"   \"Y1919_VU\"  \n",
      " [841] \"Y1920_VU\"   \"Y1921_VU\"   \"Y1922_VU\"   \"Y1923_VU\"   \"Y1924_VL\"  \n",
      " [846] \"Y1925_VL\"   \"Y1926_VL\"   \"Y1927_VL\"   \"Y1928_VL\"   \"Y1929_VL\"  \n",
      " [851] \"Y1930_VL\"   \"Y1931_VL\"   \"Y1932_VL\"   \"Y1933_VL\"   \"Y1934_VL\"  \n",
      " [856] \"Y1935_VL\"   \"Y1936_VL\"   \"Y1937_VL\"   \"Y1938_VL\"   \"Y1939_VL\"  \n",
      " [861] \"Y1940_VL\"   \"Y1941_VL\"   \"Y1942_VL\"   \"Y1943_VL\"   \"Y1944_VL\"  \n",
      " [866] \"Y1945_VL\"   \"Y1946_VL\"   \"Y1947_VL\"   \"Y2052_VU\"   \"Y2053_VU\"  \n",
      " [871] \"Y2054_VU\"   \"Y2055_VU\"   \"Y2056_VU\"   \"Y2057_VU\"   \"Y2058_VU\"  \n",
      " [876] \"Y2059_VU\"   \"Y2060_VL\"   \"Y2061_VL\"   \"Y2062_VL\"   \"Y2063_VL\"  \n",
      " [881] \"Y2064_VL\"   \"Y2065_VL\"   \"Y2066_VL\"   \"Y2067_VL\"   \"Y1429_V\"   \n",
      " [886] \"Y1430_V\"    \"Y1431_V\"    \"Y1432_V\"    \"Y1433_V\"    \"Y1434_V\"   \n",
      " [891] \"Y1435_V\"    \"Y1436_V\"    \"Y1437_V\"    \"Y1438_V\"    \"Y1439_V\"   \n",
      " [896] \"Y1440_V\"    \"Y1441_V\"    \"Y1442_V\"    \"Y1443_V\"    \"Y1444_V\"   \n",
      " [901] \"Y1445_V\"    \"Y1446_V\"    \"Y1447_V\"    \"Y1448_V\"    \"Y1449_V\"   \n",
      " [906] \"Y1450_V\"    \"Y1451_V\"    \"Y1452_V\"    \"Y1453_V\"    \"P216_V\"    \n",
      " [911] \"P217_V\"     \"P218_V\"     \"P219_V\"     \"P220_V\"     \"P221_V\"    \n",
      " [916] \"P222_V\"     \"P223_V\"     \"P1527_V\"    \"P1528_V\"    \"P1529_V\"   \n",
      " [921] \"P1530_V\"    \"P1531_V\"    \"P1532_V\"    \"P1533_V\"    \"P1534_V\"   \n",
      " [926] \"P1535_V\"    \"P1536_V\"    \"P1537_V\"    \"P1538_V\"    \"P1539_V\"   \n",
      " [931] \"P1540_V\"    \"P1541_V\"    \"P1542_V\"    \"P1543_V\"    \"P1544_V\"   \n",
      " [936] \"P1545_V\"    \"P1546_V\"    \"P1547_V\"    \"P363_V\"     \"P364_V\"    \n",
      " [941] \"P365_V\"     \"P367_V\"     \"P368_V\"     \"P369_V\"     \"P370_V\"    \n",
      " [946] \"P371_V\"     \"P372_V\"     \"P373_V\"     \"P374_V\"     \"P375_V\"    \n",
      " [951] \"P376_V\"     \"P377_V\"     \"P378_V\"     \"G1735_V\"    \"G1736_V\"   \n",
      " [956] \"G1738_V\"    \"G1739_V\"    \"G1740_V\"    \"G1741_V\"    \"G1742_V\"   \n",
      " [961] \"G1743_V\"    \"G1744_V\"    \"G1745_V\"    \"G1746_V\"    \"G1747_V\"   \n",
      " [966] \"G1748_V\"    \"G1749_V\"    \"G1750_V\"    \"G1751_V\"    \"G1752_V\"   \n",
      " [971] \"G1753_V\"    \"G1754_V\"    \"G1755_V\"    \"G1756_V\"    \"G1757_V\"   \n",
      " [976] \"G1758_V\"    \"G1759_V\"    \"G1760_V\"    \"G1761_V\"    \"G1762_V\"   \n",
      " [981] \"G1763_V\"    \"G1764_V\"    \"G1765_V\"    \"G1766_V\"    \"G1767_V\"   \n",
      " [986] \"G1768_V\"    \"G1769_V\"    \"G1770_V\"    \"G1771_V\"    \"G1772_V\"   \n",
      " [991] \"G1773_V\"    \"G1774_V\"    \"G1775_V\"    \"G1776_V\"    \"G1777_V\"   \n",
      " [996] \"G1778_V\"    \"G1779_V\"    \"G1075_VU\"   \"G1076_VU\"   \"G1077_VU\"  \n",
      "[1001] \"G1078_VU\"   \"G1079_VU\"   \"G1080_VU\"   \"G1081_VU\"   \"G1082_VU\"  \n",
      "[1006] \"G1083_VU\"   \"G1084_VU\"   \"G1085_VU\"   \"G1086_VU\"   \"G1088_VU\"  \n",
      "[1011] \"G1089_VU\"   \"G1090_VU\"   \"G1091_VU\"   \"G1092_VU\"   \"G1093_VU\"  \n",
      "[1016] \"G1094_VU\"   \"G1095_VU\"   \"G1096_VU\"   \"G1097_VU\"   \"G1098_VU\"  \n",
      "[1021] \"G1147_VL\"   \"G1148_VL\"   \"G1149_VL\"   \"G1150_VL\"   \"G1151_VL\"  \n",
      "[1026] \"G1152_VL\"   \"G1155_VL\"   \"G1156_VL\"   \"G1157_VL\"   \"G1158_VL\"  \n",
      "[1031] \"G1159_VL\"   \"G1160_VL\"   \"G1161_VL\"   \"G1163_VL\"   \"G1164_VL\"  \n",
      "[1036] \"G1165_VL\"   \"G1166_VL\"   \"G1167_VL\"   \"G1168_VL\"   \"G1169_VL\"  \n",
      "[1041] \"R224_V5\"    \"R225_V5\"    \"R226_V5\"    \"R227_V5\"    \"R228_V5\"   \n",
      "[1046] \"R229_V5\"    \"R230_V5\"    \"R231_V5\"    \"R2068_V\"    \"R2070_V\"   \n",
      "[1051] \"R2071_V\"    \"R2072_V\"    \"R2073_V\"    \"R2074_V\"    \"R2075_V\"   \n",
      "[1056] \"R2076_V\"    \"R2077_V\"    \"R2078_V\"    \"R2079_V\"    \"R2080_V\"   \n",
      "[1061] \"R2081_V\"    \"R2082_V\"    \"R2083_V\"    \"R2084_V\"    \"R2085_V\"   \n",
      "[1066] \"R2086_V\"    \"R2087_V\"    \"R2088_V\"    \"R2089_V\"    \"R2090_V\"   \n",
      "[1071] \"R2091_V\"    \"R2092_V\"    \"R2093_V\"    \"R2095_V\"    \"R2096_V\"   \n",
      "[1076] \"R2097_V\"    \"R2098_V\"    \"R2099_V\"    \"R2100_V\"    \"R2101_V\"   \n",
      "[1081] \"R2102_V\"    \"R2103_V\"    \"R2104_V\"    \"R2105_V\"    \"R2106_V\"   \n",
      "[1086] \"R2107_V\"    \"R419_V5\"    \"R420_V5\"    \"R421_V5\"    \"R424_V5\"   \n",
      "[1091] \"R425_V5\"    \"R426_V5\"    \"R427_V5\"    \"R428_V5\"    \"R429_V5\"   \n",
      "[1096] \"R430_V5\"    \"R431_V5\"    \"R432_V5\"    \"R433_V5\"    \"R434_V5\"   \n",
      "[1101] \"R435_V5\"    \"R436_V5\"    \"R437_V5\"    \"R438_V5\"    \"R439_V5\"   \n",
      "[1106] \"R440_V5\"    \"R441_V5\"    \"R442_V5\"    \"R606_V5\"    \"R607_V5\"   \n",
      "[1111] \"R608_V5\"    \"R609_V5\"    \"R610_V5\"    \"R611_V5\"    \"R612_V5\"   \n",
      "[1116] \"R613_V5\"    \"R614_V5\"    \"R615_V5\"    \"R616_V5\"    \"R617_V5\"   \n",
      "[1121] \"R618_V5\"    \"R619_V5\"    \"R620_V5\"    \"R621_V5\"    \"R622_V5\"   \n",
      "[1126] \"R623_V5\"    \"R624_V5\"    \"R625_V5\"    \"R626_V5\"    \"R627_V5\"   \n",
      "[1131] \"R628_V5\"    \"R629_V5\"    \"R630_V5\"    \"R631_V5\"    \"R632_V5\"   \n",
      "[1136] \"R633_V5\"    \"R634_V5\"    \"R635_V5\"    \"R636_V5\"    \"R637_V5\"   \n",
      "[1141] \"R638_V5\"    \"R639_V5\"    \"R640_V5\"    \"R641_V5\"    \"R642_V5\"   \n",
      "[1146] \"R643_V5\"    \"R644_V5\"    \"R645_V5\"    \"R829_V5\"    \"R830_V5\"   \n",
      "[1151] \"R831_V5\"    \"R832_V5\"    \"R833_V5\"    \"R834_V5\"    \"R835_V5\"   \n",
      "[1156] \"R836_V5\"    \"R837_V5\"    \"R838_V5\"    \"R839_V5\"    \"R840_V5\"   \n",
      "[1161] \"R841_V5\"    \"R842_V5\"    \"R843_V5\"    \"R844_V5\"    \"R291_V5\"   \n",
      "[1166] \"R292_V5\"    \"R293_V5\"    \"R294_V5\"    \"R295_V5\"    \"R296_V5\"   \n",
      "[1171] \"R297_V5\"    \"R298_V5\"    \"R305_V5\"    \"R309_V5\"    \"R314_V5\"   \n",
      "[1176] \"R315_V5\"    \"R316_V5\"    \"R317_V5\"    \"R318_V5\"    \"R319_V5\"   \n",
      "[1181] \"R320_V5\"    \"R321_V\"     \"R322_V\"     \"R323_V\"     \"R324_V\"    \n",
      "[1186] \"R325_V\"     \"R326_V\"     \"R327_V\"     \"R328_V\"     \"R329_V\"    \n",
      "[1191] \"R330_V\"     \"R331_V\"     \"R332_V\"     \"R333_V\"     \"R334_V\"    \n",
      "[1196] \"R335_V\"     \"R336_V\"     \"R337_V\"     \"R338_V\"     \"R339_V\"    \n",
      "[1201] \"R340_V\"     \"R341_V\"     \"R342_V\"     \"R343_V\"     \"R344_V5\"   \n",
      "[1206] \"R355_V\"     \"R356_V\"     \"R357_V\"     \"R358_V\"     \"R359_V\"    \n",
      "[1211] \"R360_V\"     \"R361_V\"     \"R362_V\"     \"R395n_V5\"   \"R396n_V5\"  \n",
      "[1216] \"R397n_V5\"   \"R398n_V5\"   \"R399n_V5\"   \"R400n_V5\"   \"R401n_V5\"  \n",
      "[1221] \"R402n_V5\"   \"R403n_V5\"   \"R404n_V5\"   \"R405n_V5\"   \"R406n_V5\"  \n",
      "[1226] \"R407n_V5\"   \"R408n_V5\"   \"R409n_V5\"   \"R410n_V5\"   \"R412n_V5\"  \n",
      "[1231] \"R413n_V5\"   \"R414n_V5\"   \"R415n_V5\"   \"R416n_V5\"   \"R417n_V5\"  \n",
      "[1236] \"R418n_V5\"   \"B590_V4\"    \"B591_V4\"    \"B592_V4\"    \"B593_V4\"   \n",
      "[1241] \"B594_V4\"    \"B595_V4\"    \"B597_V4\"    \"B598_V4\"    \"B599_V4\"   \n",
      "[1246] \"B600_V4\"    \"B601_V4\"    \"B602_V4\"    \"B603_V4\"    \"B604_V4\"   \n",
      "[1251] \"B605_V4\"    \"B821_V4\"    \"B822_V4\"    \"B823_V4\"    \"B824_V4\"   \n",
      "[1256] \"B825_V4\"    \"B826_V4\"    \"B827_V4\"    \"B828_V4\"    \"B80_V4\"    \n",
      "[1261] \"B81_V4\"     \"B82_V4\"     \"B83_V4\"     \"B84_V4\"     \"B85_V4\"    \n",
      "[1266] \"B86_V4\"     \"B87_V4\"     \"B160_V4\"    \"B161_V4\"    \"B162_V4\"   \n",
      "[1271] \"B164_V4\"    \"B165_V4\"    \"B166_V4\"    \"B167_V4\"    \"B379_V5\"   \n",
      "[1276] \"B380_V5\"    \"B381_V5\"    \"B382_V5\"    \"B383_V5\"    \"B384_V5\"   \n",
      "[1281] \"B385_V5\"    \"B386_V5\"    \"B387_V5\"    \"B388_V5\"    \"B389_V5\"   \n",
      "[1286] \"B390_V5\"    \"B391_V5\"    \"B392_V5\"    \"B393_V5\"    \"B394_V5\"   \n",
      "[1291] \"Q286_V4\"    \"Q287_V4\"    \"Q288_V4\"    \"Q289_V4\"    \"Q290_V4\"   \n",
      "[1296] \"Q582_V4\"    \"Q583_V4\"    \"Q584_V4\"    \"Q585_V4\"    \"Q586_V4\"   \n",
      "[1301] \"Q588_V4\"    \"Q589_V4\"    \"Q1060_V4\"   \"Q1061_V4\"   \"Q1062_V4\"  \n",
      "[1306] \"Q1063_V4\"   \"Q1064_V4\"   \"Q1065_V4\"   \"Q1066_V4\"   \"F1069_V1\"  \n",
      "[1311] \"F1191_V1\"   \"F1192_V1\"   \"F1193_V1\"   \"F1194_V1\"   \"F1195_V1\"  \n",
      "[1316] \"F1196_V1\"   \"F1197_V1\"   \"F1198_V1\"   \"F1199_V1\"   \"F1200_V1\"  \n",
      "[1321] \"F1201_V1\"   \"F1202_V1\"   \"F1203_V1\"   \"F1204_V1\"   \"F1205_V1\"  \n",
      "[1326] \"F1206_V1\"   \"F1207_V1\"   \"F1208_V1\"   \"S265_V4\"    \"S266_V4\"   \n",
      "[1331] \"S267_V4\"    \"S268_V4\"    \"S269_V4\"    \"S270_V4\"    \"S271_V4\"   \n",
      "[1336] \"S462_V4\"    \"S463_V4\"    \"S464_V4\"    \"S465_V4\"    \"S466_V4\"   \n",
      "[1341] \"S467_V4\"    \"S469_V4\"    \"S470_V4\"    \"S471_V4\"    \"S472_V4\"   \n",
      "[1346] \"S473_V4\"    \"S474_V4\"    \"S475_V4\"    \"S476_V4\"    \"S477_V4\"   \n",
      "[1351] \"S518_V4\"    \"S519_V4\"    \"S521_V4\"    \"S522_V4\"    \"S523_V4\"   \n",
      "[1356] \"S524_V4\"    \"S525_V4\"    \"S526_V4\"    \"S527_V4\"    \"S528_V4\"   \n",
      "[1361] \"S529_V4\"    \"S530_V4\"    \"S531_V4\"    \"S532_V4\"    \"S533_V4\"   \n",
      "[1366] \"S534_V4\"    \"S535_V4\"    \"S536_V4\"    \"S537_V4\"    \"S538_V4\"   \n",
      "[1371] \"S539_V4\"    \"S540_V4\"    \"S542_V4\"    \"S543_V4\"    \"S544_V4\"   \n",
      "[1376] \"S545_V4\"    \"S546_V4\"    \"S547_V4\"    \"S548_V4\"    \"S550_V4\"   \n",
      "[1381] \"S551_V4\"    \"S552_V4\"    \"S554_V4\"    \"S555_V4\"    \"S556_V4\"   \n",
      "[1386] \"S557_V4\"    \"S1054_V56\"  \"S1055_V56\"  \"S1056_V56\"  \"S1057_V56\" \n",
      "[1391] \"S1058_V56\"  \"S1059_V56\"  \"S1275_V\"    \"S1276_V\"    \"S1277_V\"   \n",
      "[1396] \"S1278_V\"    \"S1279_V\"    \"S1280_V\"    \"S1282_V\"    \"S1283_V\"   \n",
      "[1401] \"S1285_V\"    \"S1286_V\"    \"S1287_V\"    \"S1288_V\"    \"S1289_V\"   \n",
      "[1406] \"S1290_V\"    \"S1291_V\"    \"S1293_V\"    \"S1294_V\"    \"S1295_V\"   \n",
      "[1411] \"S1296_V\"    \"S1297_V\"    \"S1299_V\"    \"S1300_V\"    \"S1301_V\"   \n",
      "[1416] \"S1302_V\"    \"S1303_V\"    \"S1304_V\"    \"S1305_V\"    \"S136_V4\"   \n",
      "[1421] \"S137_V4\"    \"S138_V4\"    \"S139_V4\"    \"S140_V4\"    \"S141_V4\"   \n",
      "[1426] \"S142_V4\"    \"S143_V4\"    \"S558n_V4\"   \"S559n_V4\"   \"S560n_V4\"  \n",
      "[1431] \"S561n_V4\"   \"S562n_V4\"   \"S563n_V4\"   \"S564n_V4\"   \"S565n_V4\"  \n",
      "[1436] \"S566n_V4\"   \"S567n_V4\"   \"S568n_V4\"   \"S569n_V4\"   \"S570n_V4\"  \n",
      "[1441] \"S571n_V4\"   \"S573n_V4\"   \"S574n_V4\"   \"S575n_V4\"   \"S576n_V4\"  \n",
      "[1446] \"S577n_V4\"   \"S578n_V4\"   \"S579n_V4\"   \"S580n_V4\"   \"S581n_V4\"  \n",
      "[1451] \"Sc1853_V\"   \"Sc1854_V\"   \"Sc1855_V\"   \"Sc1856_V\"   \"Sc1857_V\"  \n",
      "[1456] \"Sc1858_V\"   \"Sc1859_V\"   \"Sc1860_V\"   \"Sc1861_V\"   \"Sc1862_V\"  \n",
      "[1461] \"Sc1863_V\"   \"Sc1864_V\"   \"Sc1865_V\"   \"Sc1866_V\"   \"Sc1867_V\"  \n",
      "[1466] \"Sc1868_V\"   \"Sc1869_V\"   \"Sc1870_V\"   \"Sc1871_V\"   \"Sc1872_V\"  \n",
      "[1471] \"Sc1873_V\"   \"Sc1874_V\"   \"Sc1875_V\"   \"Sc1876_V\"   \"Sc1877_V\"  \n",
      "[1476] \"Sc1878_V\"   \"Sc1879_V\"   \"Sc1880_V\"   \"Sc1881_V\"   \"Sc1882_V\"  \n",
      "[1481] \"Sc1883_V\"   \"Sc1884_V\"   \"Sc1885_V\"   \"Sc1886_V\"   \"Sc1887_V\"  \n",
      "[1486] \"Sc1888_V\"   \"Sc1889_V\"   \"Sc1890_V\"   \"Sc1891_V\"   \"Sc1892_V\"  \n",
      "[1491] \"Sc1893_V\"   \"Sc1894_V\"   \"Sc1895_V\"   \"Sc1896_V\"   \"Sc1897_V\"  \n",
      "[1496] \"Sc1898_V\"   \"Sc1899_V\"   \"M2032_VU\"   \"M2033_VU\"   \"M2034_VU\"  \n",
      "[1501] \"M2035_VU\"   \"M2036_VU\"   \"M2037_VU\"   \"M2038_VU\"   \"M2039_VU\"  \n",
      "[1506] \"M2040_VU\"   \"M2041_VU\"   \"M2042_VU\"   \"M2043_VU\"   \"M2044_VU\"  \n",
      "[1511] \"M2045_VU\"   \"M2046_VU\"   \"M2047_VU\"   \"M2048_VU\"   \"M2049_VU\"  \n",
      "[1516] \"M2050_VU\"   \"M2051_VU\"   \"M2108_V\"    \"M2109_V\"    \"M2110_V\"   \n",
      "[1521] \"M240_V\"     \"M241_V\"     \"M243_V\"     \"M244_V\"     \"M245_V\"    \n",
      "[1526] \"M246_V\"     \"M247_V\"     \"M646_V\"     \"M647_V\"     \"M648_V\"    \n",
      "[1531] \"M649_V\"     \"M650_V\"     \"M651_V\"     \"M652_V\"     \"M653_V\"    \n",
      "[1536] \"M782_VU\"    \"M783_VU\"    \"M784_VU\"    \"M785_VU\"    \"M786_VU\"   \n",
      "[1541] \"M787_VU\"    \"M788_VU\"    \"M789_VU\"    \"M790_VU\"    \"M791_VU\"   \n",
      "[1546] \"M792_VU\"    \"M793_VU\"    \"M794_VU\"    \"M795_VU\"    \"M796_VU\"   \n",
      "[1551] \"M797_VU\"    \"M798_VU\"    \"M799_VU\"    \"M800_VU\"    \"M801_VU\"   \n",
      "[1556] \"M802_VU\"    \"M803_VU\"    \"M804_VU\"    \"M806_VL\"    \"M807_VL\"   \n",
      "[1561] \"M808_VL\"    \"M809_VL\"    \"M810_VL\"    \"M811_VL\"    \"M812_VL\"   \n",
      "[1566] \"M814_VL\"    \"M815_VL\"    \"M816_VL\"    \"M817_VL\"    \"M818_VL\"   \n",
      "[1571] \"M819_VL\"    \"M820_VL\"    \"M915_V\"     \"M916_V\"     \"M917_V\"    \n",
      "[1576] \"M918_V\"     \"M919_V\"     \"M920_V\"     \"M921_V\"     \"M922_V\"    \n",
      "[1581] \"M923_V\"     \"M924_V\"     \"M925_V\"     \"M926_V\"     \"M927_V\"    \n",
      "[1586] \"M928_V\"     \"M929_V\"     \"M930_V\"     \"M931_V\"     \"M932_V\"    \n",
      "[1591] \"M933_V\"     \"M934_V\"     \"M935_V\"     \"M936_V\"     \"M937_V\"    \n",
      "[1596] \"M938_V\"     \"M144_V\"     \"M145_V\"     \"M146_V\"     \"M147_V\"    \n",
      "[1601] \"M148_V\"     \"M149_V\"     \"M150_V\"     \"M151_V\"     \"Tc1948_V\"  \n",
      "[1606] \"Tc1949_V\"   \"Tc1950_V\"   \"Tc1951_V\"   \"Tc1952_V\"   \"Tc1954_V\"  \n",
      "[1611] \"Tc1955_V\"   \"Tc1956_V\"   \"V735_VU\"    \"V736_VU\"    \"V737_VU\"   \n",
      "[1616] \"V738_VU\"    \"V739_VU\"    \"V740_VU\"    \"V742_VU\"    \"V743_VU\"   \n",
      "[1621] \"V744_VU\"    \"V745_VU\"    \"V746_VU\"    \"V747_VU\"    \"V748_VU\"   \n",
      "[1626] \"V749_VU\"    \"V750_VU\"    \"V751_VU\"    \"V752_VU\"    \"V753_VU\"   \n",
      "[1631] \"V754_VU\"    \"V755_VU\"    \"V756_VU\"    \"V757_VU\"    \"V759_VL\"   \n",
      "[1636] \"V760_VL\"    \"V761_VL\"    \"V762_VL\"    \"V763_VL\"    \"V764_VL\"   \n",
      "[1641] \"V765_VL\"    \"V766_VL\"    \"V767_VL\"    \"V768_VL\"    \"V769_VL\"   \n",
      "[1646] \"V770_VL\"    \"V771_VL\"    \"V772_VL\"    \"V773_VL\"    \"V774_VL\"   \n",
      "[1651] \"V775_VL\"    \"V776_VL\"    \"V777_VL\"    \"V778_VL\"    \"V779_VL\"   \n",
      "[1656] \"V781_VL\"    \"V907_VL\"    \"V908_VL\"    \"V909_VL\"    \"V910_VL*\"  \n",
      "[1661] \"V911_VL\"    \"V912_VL\"    \"V913_VL\"    \"V914_VL\"    \"V982_V\"    \n",
      "[1666] \"V983_V\"     \"V984_V\"     \"V985_V\"     \"V986_V\"     \"V987_V*\"   \n",
      "[1671] \"V988_V\"     \"V989_V\"     \"V990_V\"     \"V991_V\"     \"V992_V\"    \n",
      "[1676] \"V993_V\"     \"V994_V\"     \"V995_V\"     \"V996_V\"     \"V997_V\"    \n"
     ]
    }
   ],
   "source": [
    "# Show unique values for categorical columns in colData\n",
    "df <- as.data.frame(SummarizedExperiment::colData(tasic))\n",
    "cat_cols <- names(Filter(function(x) is.factor(x) || is.character(x) || is.logical(x), df))\n",
    "\n",
    "invisible(lapply(cat_cols, function(nm) {\n",
    "  cat(\"\\n==\", nm, \"==\\n\", sep = \"\")\n",
    "  print(unique(df[[nm]]))\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "eb002c35-037c-4f4d-8330-e53f778e8e1d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "==tissue==\n",
      "[1] \"sscortex\"       \"ca1hippocampus\"\n",
      "\n",
      "==level1class==\n",
      "[1] \"interneurons\"         \"pyramidal SS\"         \"pyramidal CA1\"       \n",
      "[4] \"oligodendrocytes\"     \"microglia\"            \"endothelial-mural\"   \n",
      "[7] \"astrocytes_ependymal\"\n",
      "\n",
      "==level2class==\n",
      " [1] \"Int10\"     \"Int6\"      \"Int9\"      \"Int2\"      \"Int4\"      \"Int1\"     \n",
      " [7] \"Int3\"      \"Int13\"     \"Int16\"     \"Int14\"     \"Int11\"     \"Int5\"     \n",
      "[13] \"Int7\"      \"Int8\"      \"Int12\"     \"Int15\"     \"(none)\"    \"S1PyrL4\"  \n",
      "[19] \"ClauPyr\"   \"S1PyrL5\"   \"S1PyrL23\"  \"S1PyrDL\"   \"S1PyrL5a\"  \"SubPyr\"   \n",
      "[25] \"CA1Pyr1\"   \"S1PyrL6b\"  \"S1PyrL6\"   \"CA1Pyr2\"   \"CA1PyrInt\" \"CA2Pyr2\"  \n",
      "[31] \"Oligo1\"    \"Oligo3\"    \"Oligo4\"    \"Oligo2\"    \"Oligo6\"    \"Oligo5\"   \n",
      "[37] \"Mgl1\"      \"Mgl2\"      \"Pvm1\"      \"Pvm2\"      \"Vsmc\"      \"Vend2\"    \n",
      "[43] \"Peric\"     \"Vend1\"     \"Astro2\"    \"Astro1\"    \"Choroid\"   \"Epend\"    \n"
     ]
    }
   ],
   "source": [
    "# Show unique values for categorical columns in colData\n",
    "df <- as.data.frame(SummarizedExperiment::colData(zeisel))\n",
    "cat_cols <- names(Filter(function(x) is.factor(x) || is.character(x) || is.logical(x), df))\n",
    "\n",
    "invisible(lapply(cat_cols, function(nm) {\n",
    "  cat(\"\\n==\", nm, \"==\\n\", sep = \"\")\n",
    "  print(unique(df[[nm]]))\n",
    "}))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "54454a5e-4a66-4e02-a2aa-b9b173aaff97",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make rds objects from MTG_SEA resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1b5245b0-6212-4bb0-ac64-75d2bde25128",
   "metadata": {},
   "outputs": [],
   "source": [
    "# /mnt/nfs/CX000008_DS1/projects/btanasa/brain_refs_MTG_SE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "388db43c-56a6-40fb-b16e-3d8ca14e0fe5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# working with a file from : /mnt/nfs/CX000008_DS1/projects/btanasa/brain_refs_MTG_SE\n",
    "\n",
    "# setwd(\"/mnt/nfs/CX000008_DS1/projects/btanasa/brain_refs_MTG_SEA\")\n",
    "# dir=\"./\"\n",
    "\n",
    "# suppressPackageStartupMessages({\n",
    "#  library(data.table)\n",
    "#  library(Matrix)\n",
    "#  library(SingleCellExperiment)\n",
    "#  library(SummarizedExperiment)\n",
    "# })\n",
    "\n",
    "# mfile <- file.path(dir, \"matrix.csv\")\n",
    "# meta  <- file.path(dir, \"metadata.csv\")\n",
    "\n",
    "# dt <- fread(mfile)   \n",
    "# head(dt[1:10,1:10])\n",
    "\n",
    "# md <- fread(meta)\n",
    "# colnames(md)\n",
    "# head(md,2)\n",
    "\n",
    "# unique(md$cluster_label)\n",
    "# unique(md$class_label)\n",
    "# unique(md$subclass_label)\n",
    "\n",
    "# cell_ids <- as.character(dt[[\"sample_name\"]])\n",
    "# gene_ids <- as.character(names(dt)[-1])\n",
    "\n",
    "# fast path (dense -> transpose -> sparse)\n",
    "# counts <- try({\n",
    "#  M <- as.matrix(dt[, -1, with = FALSE])   # cells x genes\n",
    "#  mode(M) <- \"numeric\"\n",
    "#  M <- t(M)                                # genes x cells\n",
    "#  colnames(M) <- cell_ids\n",
    "#  rownames(M) <- gene_ids\n",
    "#  Matrix(M, sparse = TRUE)\n",
    "# }, silent = TRUE)\n",
    "\n",
    "# o <- match(colnames(counts), md$sample_name)\n",
    "\n",
    "# dim(counts)\n",
    "# head(o)\n",
    "\n",
    "# o <- match(colnames(counts), md$sample_name)\n",
    "# if (anyNA(o))\n",
    "#  stop(\"Some cells in matrix.csv are missing from metadata.csv via 'sample_name'.\")\n",
    "# md <- as.data.frame(md[o, , drop = FALSE])\n",
    "\n",
    "#  Build SCE and add logcounts\n",
    "# sce <- SingleCellExperiment(\n",
    "#  assays  = list(counts = counts),\n",
    "#  colData = S4Vectors::DataFrame(md)\n",
    "# )\n",
    "\n",
    "# sce <- logNormCounts(sce)\n",
    "\n",
    "# sce                             # class, dims, assays, reduced dims, altExps\n",
    "# class(sce)\n",
    "# dim(sce)                        # genes x cells\n",
    "# assayNames(sce)                 # e.g., \"counts\", \"logcounts\"\n",
    "# head(rownames(sce), 5)          # first genes\n",
    "# head(colnames(sce), 5)          # first cells (if present)\n",
    "\n",
    "# names(rowData(sce))             # feature metadata fields\n",
    "# names(colData(sce))             # cell metadata fields\n",
    "# head(as.data.frame(colData(sce))[ , 1:min(10, ncol(colData(sce))), drop=FALSE])\n",
    "\n",
    "# reducedDimNames(sce)            # embeddings stored (e.g., PCA/TSNE/UMAP)\n",
    "# altExpNames(sce)                # alternative experiments (e.g., ERCC, repeats)\n",
    "# metadata(sce)                   # list of extra metadata (taxonomy, dendrogram, etc.)\n",
    "\n",
    "# --- Basic counts/logcounts checks ---\n",
    "# has_counts   <- \"counts\"    %in% assayNames(sce)\n",
    "# has_logcounts<- \"logcounts\" %in% assayNames(sce)\n",
    "# cat(\"Has counts:\", has_counts, \" | Has logcounts:\", has_logcounts, \"\\n\")\n",
    "\n",
    "# saveRDS(sce, file.path(dir, \"brain_refs_MTG_SEA.from.matrix.metadata.sce.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a742662-c5d5-4c48-a5b2-24af96355950",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "97026102-a8a5-4294-b02e-89289595d118",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'SingleCellExperiment'"
      ],
      "text/latex": [
       "'SingleCellExperiment'"
      ],
      "text/markdown": [
       "'SingleCellExperiment'"
      ],
      "text/plain": [
       "[1] \"SingleCellExperiment\"\n",
       "attr(,\"package\")\n",
       "[1] \"SingleCellExperiment\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'sample_name'</li><li>'exp_component_name'</li><li>'cluster_label'</li><li>'cluster_color'</li><li>'cluster_order'</li><li>'class_label'</li><li>'class_color'</li><li>'class_order'</li><li>'subclass_label'</li><li>'subclass_color'</li><li>'subclass_order'</li><li>'donor_sex_label'</li><li>'donor_sex_color'</li><li>'donor_sex_order'</li><li>'region_label'</li><li>'region_color'</li><li>'region_order'</li><li>'cortical_layer_label'</li><li>'cortical_layer_color'</li><li>'cortical_layer_order'</li><li>'cell_type_accession_label'</li><li>'cell_type_accession_color'</li><li>'cell_type_accession_order'</li><li>'cell_type_alias_label'</li><li>'cell_type_alias_color'</li><li>'cell_type_alias_order'</li><li>'cell_type_alt_alias_label'</li><li>'cell_type_alt_alias_color'</li><li>'cell_type_alt_alias_order'</li><li>'cell_type_designation_label'</li><li>'cell_type_designation_color'</li><li>'cell_type_designation_order'</li><li>'external_donor_name_label'</li><li>'external_donor_name_color'</li><li>'external_donor_name_order'</li><li>'specimen_type'</li><li>'full_genotype_label'</li><li>'outlier_call'</li><li>'outlier_type'</li><li>'sizeFactor'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'sample\\_name'\n",
       "\\item 'exp\\_component\\_name'\n",
       "\\item 'cluster\\_label'\n",
       "\\item 'cluster\\_color'\n",
       "\\item 'cluster\\_order'\n",
       "\\item 'class\\_label'\n",
       "\\item 'class\\_color'\n",
       "\\item 'class\\_order'\n",
       "\\item 'subclass\\_label'\n",
       "\\item 'subclass\\_color'\n",
       "\\item 'subclass\\_order'\n",
       "\\item 'donor\\_sex\\_label'\n",
       "\\item 'donor\\_sex\\_color'\n",
       "\\item 'donor\\_sex\\_order'\n",
       "\\item 'region\\_label'\n",
       "\\item 'region\\_color'\n",
       "\\item 'region\\_order'\n",
       "\\item 'cortical\\_layer\\_label'\n",
       "\\item 'cortical\\_layer\\_color'\n",
       "\\item 'cortical\\_layer\\_order'\n",
       "\\item 'cell\\_type\\_accession\\_label'\n",
       "\\item 'cell\\_type\\_accession\\_color'\n",
       "\\item 'cell\\_type\\_accession\\_order'\n",
       "\\item 'cell\\_type\\_alias\\_label'\n",
       "\\item 'cell\\_type\\_alias\\_color'\n",
       "\\item 'cell\\_type\\_alias\\_order'\n",
       "\\item 'cell\\_type\\_alt\\_alias\\_label'\n",
       "\\item 'cell\\_type\\_alt\\_alias\\_color'\n",
       "\\item 'cell\\_type\\_alt\\_alias\\_order'\n",
       "\\item 'cell\\_type\\_designation\\_label'\n",
       "\\item 'cell\\_type\\_designation\\_color'\n",
       "\\item 'cell\\_type\\_designation\\_order'\n",
       "\\item 'external\\_donor\\_name\\_label'\n",
       "\\item 'external\\_donor\\_name\\_color'\n",
       "\\item 'external\\_donor\\_name\\_order'\n",
       "\\item 'specimen\\_type'\n",
       "\\item 'full\\_genotype\\_label'\n",
       "\\item 'outlier\\_call'\n",
       "\\item 'outlier\\_type'\n",
       "\\item 'sizeFactor'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'sample_name'\n",
       "2. 'exp_component_name'\n",
       "3. 'cluster_label'\n",
       "4. 'cluster_color'\n",
       "5. 'cluster_order'\n",
       "6. 'class_label'\n",
       "7. 'class_color'\n",
       "8. 'class_order'\n",
       "9. 'subclass_label'\n",
       "10. 'subclass_color'\n",
       "11. 'subclass_order'\n",
       "12. 'donor_sex_label'\n",
       "13. 'donor_sex_color'\n",
       "14. 'donor_sex_order'\n",
       "15. 'region_label'\n",
       "16. 'region_color'\n",
       "17. 'region_order'\n",
       "18. 'cortical_layer_label'\n",
       "19. 'cortical_layer_color'\n",
       "20. 'cortical_layer_order'\n",
       "21. 'cell_type_accession_label'\n",
       "22. 'cell_type_accession_color'\n",
       "23. 'cell_type_accession_order'\n",
       "24. 'cell_type_alias_label'\n",
       "25. 'cell_type_alias_color'\n",
       "26. 'cell_type_alias_order'\n",
       "27. 'cell_type_alt_alias_label'\n",
       "28. 'cell_type_alt_alias_color'\n",
       "29. 'cell_type_alt_alias_order'\n",
       "30. 'cell_type_designation_label'\n",
       "31. 'cell_type_designation_color'\n",
       "32. 'cell_type_designation_order'\n",
       "33. 'external_donor_name_label'\n",
       "34. 'external_donor_name_color'\n",
       "35. 'external_donor_name_order'\n",
       "36. 'specimen_type'\n",
       "37. 'full_genotype_label'\n",
       "38. 'outlier_call'\n",
       "39. 'outlier_type'\n",
       "40. 'sizeFactor'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"sample_name\"                 \"exp_component_name\"         \n",
       " [3] \"cluster_label\"               \"cluster_color\"              \n",
       " [5] \"cluster_order\"               \"class_label\"                \n",
       " [7] \"class_color\"                 \"class_order\"                \n",
       " [9] \"subclass_label\"              \"subclass_color\"             \n",
       "[11] \"subclass_order\"              \"donor_sex_label\"            \n",
       "[13] \"donor_sex_color\"             \"donor_sex_order\"            \n",
       "[15] \"region_label\"                \"region_color\"               \n",
       "[17] \"region_order\"                \"cortical_layer_label\"       \n",
       "[19] \"cortical_layer_color\"        \"cortical_layer_order\"       \n",
       "[21] \"cell_type_accession_label\"   \"cell_type_accession_color\"  \n",
       "[23] \"cell_type_accession_order\"   \"cell_type_alias_label\"      \n",
       "[25] \"cell_type_alias_color\"       \"cell_type_alias_order\"      \n",
       "[27] \"cell_type_alt_alias_label\"   \"cell_type_alt_alias_color\"  \n",
       "[29] \"cell_type_alt_alias_order\"   \"cell_type_designation_label\"\n",
       "[31] \"cell_type_designation_color\" \"cell_type_designation_order\"\n",
       "[33] \"external_donor_name_label\"   \"external_donor_name_color\"  \n",
       "[35] \"external_donor_name_order\"   \"specimen_type\"              \n",
       "[37] \"full_genotype_label\"         \"outlier_call\"               \n",
       "[39] \"outlier_type\"                \"sizeFactor\"                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster_label: 127 unique\n",
      "  [1] \"Inh L1-2 SST CCNJL\"          \"Exc L5-6 FEZF2 IFNG-AS1\"    \n",
      "  [3] \"Exc L3-5 RORB LINC01202\"     \"Exc L2 LINC00507 GLRA3\"     \n",
      "  [5] \"Oligo L2-6 OPALIN FTH1P3\"    \"Exc L5-6 FEZF2 C9orf135-AS1\"\n",
      "  [7] \"Exc L3-5 FEZF2 ASGR2\"        \"Exc L3-5 RORB LNX2\"         \n",
      "  [9] \"Exc L5 THEMIS RGPD6\"         \"Exc L3 LAMP5 CARM1P1\"       \n",
      " [11] \"Exc L2-3 RORB CCDC68\"        \"Inh L5 SST RPL35AP11\"       \n",
      " [13] \"Exc L3 RORB OTOGL\"           \"Exc L3-5 RORB RPRM\"         \n",
      " [15] \"Exc L6 FEZF2 KLK7\"           \"Inh L1-6 LAMP5 NES\"         \n",
      " [17] \"Exc L6 THEMIS LINC00343\"     \"Inh L5-6 PVALB MEPE\"        \n",
      " [19] \"Inh L1-3 VIP CHRNA2\"         \"Inh L3-5 SST GGTLC3\"        \n",
      " [21] \"Exc L5 THEMIS VILL\"          \"Exc L3 THEMIS ENPEP\"        \n",
      " [23] \"Oligo L3-6 OPALIN ENPP6\"     \"Inh L2-5 PVALB RPH3AL\"      \n",
      " [25] \"Inh L1-3 VIP HSPB6\"          \"Exc L5 FEZF2 CSN1S1\"        \n",
      " [27] \"Inh L5-6 PVALB FAM150B\"      \"Exc L5 THEMIS SLC22A18\"     \n",
      " [29] \"Inh L3-5 VIP TAC3\"           \"Exc L5 FEZF2 NREP-AS1\"      \n",
      " [31] \"Inh L5 PVALB LRIG3\"          \"Inh L5-6 PVALB SST CRHR2\"   \n",
      " [33] \"Inh L1-2 SST PRRT4\"          \"Exc L3-5 RORB TNNT2\"        \n",
      " [35] \"Exc L6 THEMIS SNTG2\"         \"Exc L5 FEZF2 PKD2L1\"        \n",
      " [37] \"Inh L3-5 PVALB ISG20\"        \"Exc L2-3 RORB PTPN3\"        \n",
      " [39] \"Inh L1-3 SST FAM20A\"         \"Inh L5-6 VIP COL4A3\"        \n",
      " [41] \"Inh L1-5 VIP SMOC1\"          \"Inh L3-5 VIP HS3ST3A1\"      \n",
      " [43] \"Exc L6 FEZF2 FFAR4\"          \"Inh L1-5 VIP LINC01013\"     \n",
      " [45] \"Inh L5-6 SST KLHL1\"          \"Inh L1-3 VIP CBLN1\"         \n",
      " [47] \"Exc L5-6 FEZF2 OR1L8\"        \"Exc L3-5 RORB LAMA4\"        \n",
      " [49] \"Inh L2-3 SST NMU\"            \"Oligo L2-6 OPALIN MAP6D1\"   \n",
      " [51] \"Inh L5-6 PVALB ZFPM2-AS1\"    \"Exc L6 FEZF2 PROKR2\"        \n",
      " [53] \"Astro L1-6 FGFR3 PLCG1\"      \"Inh L5-6 LAMP5 CRABP1\"      \n",
      " [55] \"Exc L5-6 FEZF2 FILIP1L\"      \"Inh L3 PVALB SAMD13\"        \n",
      " [57] \"Inh L3-6 VIP ZIM2-AS1\"       \"Micro L1-6 TYROBP CD74\"     \n",
      " [59] \"Exc L2 LINC00507 ATP7B\"      \"Inh L5-6 SST ISX\"           \n",
      " [61] \"Exc L2-3 LINC00507 DSG3\"     \"Exc L6 FEZF2 POGK\"          \n",
      " [63] \"Inh L6 SST TH\"               \"Exc L5-6 THEMIS TNFAIP6\"    \n",
      " [65] \"Exc L5-6 THEMIS SMYD1\"       \"Inh L3-6 VIP UG0898H09\"     \n",
      " [67] \"Inh L5-6 PVALB GAPDHP60\"     \"Inh L5-6 SST FBN2\"          \n",
      " [69] \"Inh L2 VIP SLC6A16\"          \"Inh L3-5 VIP IGDCC3\"        \n",
      " [71] \"Inh L5-6 SST C4orf26\"        \"Exc L6 THEMIS SLN\"          \n",
      " [73] \"Inh L5-6 SST PAWR\"           \"Exc L2-3 RORB RTKN2\"        \n",
      " [75] \"Inh L5-6 SST BEAN1\"          \"Exc L5-6 FEZF2 LPO\"         \n",
      " [77] \"Exc L5 RORB MED8\"            \"OPC L1-6 PDGFRA COL20A1\"    \n",
      " [79] \"Inh L1-6 LAMP5 AARD\"         \"Exc L5 THEMIS LINC01116\"    \n",
      " [81] \"Inh L2-5 PVALB HHIPL1\"       \"Inh L2 PVALB FRZB\"          \n",
      " [83] \"Inh L3-5 SST CDH3\"           \"Inh L1-2 VIP EXPH5\"         \n",
      " [85] \"Exc L5 FEZF2 RNF144A-AS1\"    \"Exc L5 THEMIS FGF10\"        \n",
      " [87] \"Inh L1-2 VIP PTGER3\"         \"Inh L1-6 PVALB COL15A1\"     \n",
      " [89] \"Inh L1-6 LAMP5 CA1\"          \"Inh L3-5 SST OR5AH1P\"       \n",
      " [91] \"Inh L5-6 SST DNAJC14\"        \"Inh L1-2 SST CLIC6\"         \n",
      " [93] \"Inh L5-6 PVALB KCNIP2\"       \"Inh L1-5 VIP PHLDB3\"        \n",
      " [95] \"Inh L5-6 SST PIK3CD\"         \"Exc L5-6 FEZF2 CFTR\"        \n",
      " [97] \"Inh L1 PVALB SST ASIC4\"      \"Exc L3-5 FEZF2 LINC01107\"   \n",
      " [99] \"Inh L2-5 VIP SOX11\"          \"Inh L2-5 VIP BSPRY\"         \n",
      "[101] \"Inh L1-5 VIP CD27-AS1\"       \"Astro L1-6 FGFR3 AQP1\"      \n",
      "[103] \"Inh L1-3 VIP FNDC1\"          \"Inh L1 SST DEFB108B\"        \n",
      "[105] \"Inh L1 LAMP5 RAB11FIP1\"      \"Astro L1 FGFR3 SERPINI2\"    \n",
      "[107] \"Oligo L5-6 OPALIN LDLRAP1\"   \"Exc L2 LAMP5 KCNG3\"         \n",
      "[109] \"Inh L1 PAX6 CHRFAM7A\"        \"Endo L2-5 NOSTRIN SRGN\"     \n",
      "[111] \"Inh L1-2 PVALB CDK20\"        \"Inh L1-6 SST NPY\"           \n",
      "[113] \"Exc L6 FEZF2 PDYN\"           \"Exc L5-6 FEZF2 SH2D1B\"      \n",
      "[115] \"Inh L1-6 VIP SLC7A6OS\"       \"Inh L1 LAMP5 PVRL2\"         \n",
      "[117] \"Inh L1-2 VIP SCML4\"          \"Inh L1 VIP KLHDC8B\"         \n",
      "[119] \"Inh L3-6 PAX6 LINC01497\"     \"Inh L2 PAX6 FREM2\"          \n",
      "[121] \"Inh L1 PAX6 MIR101-1\"        \"Inh L1 LAMP5 NMBR\"          \n",
      "[123] \"VLMC L1-5 PDGFRA COLEC12\"    \"Inh L1 LAMP5 BMP2\"          \n",
      "[125] \"Inh L1-2 VIP WNT4\"           \"Inh L1-2 VIP HTR3A\"         \n",
      "[127] \"Inh L1 SST P4HA3\"           \n",
      "class_label: 3 unique\n",
      "[1] \"GABAergic\"     \"Glutamatergic\" \"Non-Neuronal\" \n",
      "subclass_label: 20 unique\n",
      " [1] \"Sst\"        \"L5/6 NP\"    \"L5 IT\"      \"L2/3 IT\"    \"Oligo\"     \n",
      " [6] \"L6 CT\"      \"L5 ET\"      \"L6b\"        \"Lamp5\"      \"L6 IT\"     \n",
      "[11] \"Pvalb\"      \"Vip\"        \"Astro\"      \"Micro-PVM\"  \"L6 IT Car3\"\n",
      "[16] \"OPC\"        \"Sncg\"       \"Endo\"       \"Sst Chodl\"  \"VLMC\"      \n"
     ]
    }
   ],
   "source": [
    "mtgsea <- \"/mnt/nfs/CX000008_DS1/projects/btanasa/brain_refs_MTG_SEA/brain_refs_MTG_SEA.from.matrix.metadata.sce.rds\"\n",
    "mtg_sea <- readRDS(mtgsea)\n",
    "# Inspect its structure\n",
    "class(mtg_sea)\n",
    "# str(mtg_sea)\n",
    "\n",
    "# .$ cluster_label              : chr [1:76533] \"Inh L1-2 SST CCNJL\" \"Exc L5-6 FEZF2 IFNG-AS1\" \"Exc L3-5 RORB LINC01202\" \"Exc L2 LINC00507 GLRA3\" ...\n",
    "# .$ class_label                : chr [1:76533] \"GABAergic\" \"Glutamatergic\" \"Glutamatergic\" \"Glutamatergic\" ...\n",
    "# .$ subclass_label             : chr [1:76533] \"Sst\" \"L5/6 NP\" \"L5 IT\" \"L2/3 IT\" ...\n",
    "\n",
    "# Check available columns in colData\n",
    "colnames(colData(mtg_sea))\n",
    "\n",
    "# Get unique values for each label\n",
    "unique_clusters    <- unique(mtg_sea$cluster_label)\n",
    "unique_classes     <- unique(mtg_sea$class_label)\n",
    "unique_subclasses  <- unique(mtg_sea$subclass_label)\n",
    "\n",
    "# Print counts + first few\n",
    "cat(\"cluster_label:\", length(unique_clusters), \"unique\\n\")\n",
    "print(unique_clusters)\n",
    "\n",
    "cat(\"class_label:\", length(unique_classes), \"unique\\n\")\n",
    "print(unique_classes)\n",
    "\n",
    "cat(\"subclass_label:\", length(unique_subclasses), \"unique\\n\")\n",
    "print(unique_subclasses)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dd6b5480-d636-463f-b7ee-a4bba592bd57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'SingleCellExperiment'"
      ],
      "text/latex": [
       "'SingleCellExperiment'"
      ],
      "text/markdown": [
       "'SingleCellExperiment'"
      ],
      "text/plain": [
       "[1] \"SingleCellExperiment\"\n",
       "attr(,\"package\")\n",
       "[1] \"SingleCellExperiment\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Object class: SingleCellExperiment \n",
      "Total genes: 50281 \n",
      "Ensembl (ENSG) IDs: 0 \n",
      "Other (likely gene symbols): 50281 \n",
      "\n",
      "Example Ensembl IDs:\n",
      "character(0)\n",
      "\n",
      "Example non-Ensembl IDs:\n",
      "[1] \"DDX11L1\"      \"WASH7P\"       \"MIR6859-1\"    \"MIR1302-2\"    \"FAM138A\"     \n",
      "[6] \"LOC105379212\"\n"
     ]
    }
   ],
   "source": [
    "# Inspect the object type\n",
    "class(mtg_sea)\n",
    "\n",
    "# Extract rownames (gene IDs)\n",
    "ids <- sub(\"\\\\.\\\\d+$\", \"\", rownames(mtg_sea))   # strip version suffixes like .12\n",
    "\n",
    "# Flag Ensembl IDs\n",
    "is_ensg <- grepl(\"^ENSG\", ids)\n",
    "\n",
    "# Count totals\n",
    "n_total  <- length(ids)\n",
    "n_ensg   <- sum(is_ensg)\n",
    "n_symbol <- n_total - n_ensg\n",
    "\n",
    "cat(\"Object class:\", class(mtg_sea), \"\\n\")\n",
    "cat(\"Total genes:\", n_total, \"\\n\")\n",
    "cat(\"Ensembl (ENSG) IDs:\", n_ensg, \"\\n\")\n",
    "cat(\"Other (likely gene symbols):\", n_symbol, \"\\n\")\n",
    "\n",
    "# Peek at examples\n",
    "cat(\"\\nExample Ensembl IDs:\\n\")\n",
    "print(head(ids[is_ensg]))\n",
    "\n",
    "cat(\"\\nExample non-Ensembl IDs:\\n\")\n",
    "print(head(ids[!is_ensg]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e08a5549-7804-45e8-ad86-06c0ad1e1ed2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'SingleCellExperiment'"
      ],
      "text/latex": [
       "'SingleCellExperiment'"
      ],
      "text/markdown": [
       "'SingleCellExperiment'"
      ],
      "text/plain": [
       "[1] \"SingleCellExperiment\"\n",
       "attr(,\"package\")\n",
       "[1] \"SingleCellExperiment\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'orig.ident'</li><li>'nCount_RNA'</li><li>'nFeature_RNA'</li><li>'specimen_name'</li><li>'sample_name'</li><li>'donor_sex_label'</li><li>'external_donor_name_label'</li><li>'species_label'</li><li>'age_label'</li><li>'region_label'</li><li>'cortical_layer_label'</li><li>'full_genotype_label'</li><li>'QCpass'</li><li>'cluster_label'</li><li>'cluster_confidence'</li><li>'subclass_label'</li><li>'subclass_confidence'</li><li>'class_label'</li><li>'class_confidence'</li><li>'GA_QCpass'</li><li>'GA_cluster_label'</li><li>'GA_subclass_label'</li><li>'GA_neighborhood_label'</li><li>'CA_QCpass'</li><li>'CA_cluster_label'</li><li>'CA_subclass_label'</li><li>'CA_neighborhood_label'</li><li>'cluster_color'</li><li>'cluster_order'</li><li>'subclass_color'</li><li>'subclass_order'</li><li>'class_color'</li><li>'class_order'</li><li>'GA_cluster_color'</li><li>'GA_cluster_order'</li><li>'GA_subclass_color'</li><li>'GA_subclass_order'</li><li>'CA_cluster_color'</li><li>'CA_cluster_order'</li><li>'CA_subclass_color'</li><li>'CA_subclass_order'</li><li>'cell_type_accession_label'</li><li>'ident'</li><li>'seurat_clusters'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'orig.ident'\n",
       "\\item 'nCount\\_RNA'\n",
       "\\item 'nFeature\\_RNA'\n",
       "\\item 'specimen\\_name'\n",
       "\\item 'sample\\_name'\n",
       "\\item 'donor\\_sex\\_label'\n",
       "\\item 'external\\_donor\\_name\\_label'\n",
       "\\item 'species\\_label'\n",
       "\\item 'age\\_label'\n",
       "\\item 'region\\_label'\n",
       "\\item 'cortical\\_layer\\_label'\n",
       "\\item 'full\\_genotype\\_label'\n",
       "\\item 'QCpass'\n",
       "\\item 'cluster\\_label'\n",
       "\\item 'cluster\\_confidence'\n",
       "\\item 'subclass\\_label'\n",
       "\\item 'subclass\\_confidence'\n",
       "\\item 'class\\_label'\n",
       "\\item 'class\\_confidence'\n",
       "\\item 'GA\\_QCpass'\n",
       "\\item 'GA\\_cluster\\_label'\n",
       "\\item 'GA\\_subclass\\_label'\n",
       "\\item 'GA\\_neighborhood\\_label'\n",
       "\\item 'CA\\_QCpass'\n",
       "\\item 'CA\\_cluster\\_label'\n",
       "\\item 'CA\\_subclass\\_label'\n",
       "\\item 'CA\\_neighborhood\\_label'\n",
       "\\item 'cluster\\_color'\n",
       "\\item 'cluster\\_order'\n",
       "\\item 'subclass\\_color'\n",
       "\\item 'subclass\\_order'\n",
       "\\item 'class\\_color'\n",
       "\\item 'class\\_order'\n",
       "\\item 'GA\\_cluster\\_color'\n",
       "\\item 'GA\\_cluster\\_order'\n",
       "\\item 'GA\\_subclass\\_color'\n",
       "\\item 'GA\\_subclass\\_order'\n",
       "\\item 'CA\\_cluster\\_color'\n",
       "\\item 'CA\\_cluster\\_order'\n",
       "\\item 'CA\\_subclass\\_color'\n",
       "\\item 'CA\\_subclass\\_order'\n",
       "\\item 'cell\\_type\\_accession\\_label'\n",
       "\\item 'ident'\n",
       "\\item 'seurat\\_clusters'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'orig.ident'\n",
       "2. 'nCount_RNA'\n",
       "3. 'nFeature_RNA'\n",
       "4. 'specimen_name'\n",
       "5. 'sample_name'\n",
       "6. 'donor_sex_label'\n",
       "7. 'external_donor_name_label'\n",
       "8. 'species_label'\n",
       "9. 'age_label'\n",
       "10. 'region_label'\n",
       "11. 'cortical_layer_label'\n",
       "12. 'full_genotype_label'\n",
       "13. 'QCpass'\n",
       "14. 'cluster_label'\n",
       "15. 'cluster_confidence'\n",
       "16. 'subclass_label'\n",
       "17. 'subclass_confidence'\n",
       "18. 'class_label'\n",
       "19. 'class_confidence'\n",
       "20. 'GA_QCpass'\n",
       "21. 'GA_cluster_label'\n",
       "22. 'GA_subclass_label'\n",
       "23. 'GA_neighborhood_label'\n",
       "24. 'CA_QCpass'\n",
       "25. 'CA_cluster_label'\n",
       "26. 'CA_subclass_label'\n",
       "27. 'CA_neighborhood_label'\n",
       "28. 'cluster_color'\n",
       "29. 'cluster_order'\n",
       "30. 'subclass_color'\n",
       "31. 'subclass_order'\n",
       "32. 'class_color'\n",
       "33. 'class_order'\n",
       "34. 'GA_cluster_color'\n",
       "35. 'GA_cluster_order'\n",
       "36. 'GA_subclass_color'\n",
       "37. 'GA_subclass_order'\n",
       "38. 'CA_cluster_color'\n",
       "39. 'CA_cluster_order'\n",
       "40. 'CA_subclass_color'\n",
       "41. 'CA_subclass_order'\n",
       "42. 'cell_type_accession_label'\n",
       "43. 'ident'\n",
       "44. 'seurat_clusters'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"orig.ident\"                \"nCount_RNA\"               \n",
       " [3] \"nFeature_RNA\"              \"specimen_name\"            \n",
       " [5] \"sample_name\"               \"donor_sex_label\"          \n",
       " [7] \"external_donor_name_label\" \"species_label\"            \n",
       " [9] \"age_label\"                 \"region_label\"             \n",
       "[11] \"cortical_layer_label\"      \"full_genotype_label\"      \n",
       "[13] \"QCpass\"                    \"cluster_label\"            \n",
       "[15] \"cluster_confidence\"        \"subclass_label\"           \n",
       "[17] \"subclass_confidence\"       \"class_label\"              \n",
       "[19] \"class_confidence\"          \"GA_QCpass\"                \n",
       "[21] \"GA_cluster_label\"          \"GA_subclass_label\"        \n",
       "[23] \"GA_neighborhood_label\"     \"CA_QCpass\"                \n",
       "[25] \"CA_cluster_label\"          \"CA_subclass_label\"        \n",
       "[27] \"CA_neighborhood_label\"     \"cluster_color\"            \n",
       "[29] \"cluster_order\"             \"subclass_color\"           \n",
       "[31] \"subclass_order\"            \"class_color\"              \n",
       "[33] \"class_order\"               \"GA_cluster_color\"         \n",
       "[35] \"GA_cluster_order\"          \"GA_subclass_color\"        \n",
       "[37] \"GA_subclass_order\"         \"CA_cluster_color\"         \n",
       "[39] \"CA_cluster_order\"          \"CA_subclass_color\"        \n",
       "[41] \"CA_subclass_order\"         \"cell_type_accession_label\"\n",
       "[43] \"ident\"                     \"seurat_clusters\"          "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cluster_label: 128 unique\n",
      "  [1] \"Pax6_1\"       NA             \"L5/6 NP_1\"    \"L5 IT_7\"      \"L6 CT_2\"     \n",
      "  [6] \"L4 IT_2\"      \"Astro_1\"      \"L5 IT_2\"      \"L2/3 IT_5\"    \"Vip_9\"       \n",
      " [11] \"L6 CT_1\"      \"L2/3 IT_1\"    \"L5 IT_1\"      \"Sst Chodl_1\"  \"L6 IT Car3_1\"\n",
      " [16] \"L6 IT_2\"      \"Astro_2\"      \"Sncg_4\"       \"Pvalb_13\"     \"L4 IT_3\"     \n",
      " [21] \"Vip_1\"        \"L2/3 IT_13\"   \"Oligo_4\"      \"L2/3 IT_6\"    \"L2/3 IT_10\"  \n",
      " [26] \"Pvalb_15\"     \"Sst_19\"       \"Vip_19\"       \"Sst_22\"       \"L2/3 IT_8\"   \n",
      " [31] \"Sst_9\"        \"Vip_15\"       \"Sncg_1\"       \"VLMC_1\"       \"L2/3 IT_12\"  \n",
      " [36] \"L4 IT_1\"      \"Vip_18\"       \"L5 IT_5\"      \"L2/3 IT_3\"    \"Oligo_1\"     \n",
      " [41] \"L6 IT Car3_2\" \"Vip_4\"        \"Lamp5_5\"      \"Micro-PVM_2\"  \"OPC_2\"       \n",
      " [46] \"L4 IT_4\"      \"Lamp5_3\"      \"Pvalb_6\"      \"Vip_11\"       \"Oligo_2\"     \n",
      " [51] \"Lamp5_2\"      \"L6b_6\"        \"Sst_11\"       \"Lamp5_Lhx6_1\" \"Endo_1\"      \n",
      " [56] \"Lamp5_6\"      \"L6 IT Car3_3\" \"L6b_3\"        \"L5 IT_3\"      \"L5/6 NP_2\"   \n",
      " [61] \"L6b_1\"        \"Sst_13\"       \"Vip_13\"       \"Sncg_8\"       \"Vip_2\"       \n",
      " [66] \"L5 IT_6\"      \"L6b_4\"        \"L2/3 IT_7\"    \"Pvalb_3\"      \"Vip_23\"      \n",
      " [71] \"Pvalb_14\"     \"Sst_3\"        \"Sst_2\"        \"Sncg_6\"       \"L5 ET_2\"     \n",
      " [76] \"Sst_25\"       \"Vip_6\"        \"Sncg_5\"       \"Micro-PVM_1\"  \"Vip_14\"      \n",
      " [81] \"Sst_5\"        \"Pvalb_1\"      \"L6 IT_1\"      \"Sst_4\"        \"Pvalb_2\"     \n",
      " [86] \"Sst_12\"       \"Vip_12\"       \"Sst_20\"       \"Pvalb_9\"      \"Pvalb_5\"     \n",
      " [91] \"Chandelier_1\" \"VLMC_2\"       \"Pvalb_7\"      \"L6 CT_3\"      \"Pax6_3\"      \n",
      " [96] \"Astro_3\"      \"Vip_21\"       \"Sst_10\"       \"Vip_5\"        \"Lamp5_4\"     \n",
      "[101] \"L2/3 IT_2\"    \"Sst_23\"       \"Pvalb_10\"     \"Sst_7\"        \"Sncg_2\"      \n",
      "[106] \"OPC_1\"        \"Sncg_3\"       \"Lamp5_1\"      \"Astro_5\"      \"Pvalb_12\"    \n",
      "[111] \"Pvalb_8\"      \"Oligo_3\"      \"L6 CT_4\"      \"L5 ET_1\"      \"L5/6 NP_4\"   \n",
      "[116] \"Chandelier_2\" \"Vip_16\"       \"Sst_1\"        \"L5/6 NP_6\"    \"Pax6_2\"      \n",
      "[121] \"Pax6_4\"       \"L6b_2\"        \"Astro_4\"      \"L6b_5\"        \"Endo_2\"      \n",
      "[126] \"L5/6 NP_3\"    \"Sst Chodl_2\"  \"Endo_3\"      \n",
      "class_label: 4 unique\n",
      "[1] \"Neuronal: GABAergic\"         \"NA\"                         \n",
      "[3] \"Neuronal: Glutamatergic\"     \"Non-neuronal and Non-neural\"\n",
      "subclass_label: 25 unique\n",
      " [1] \"Pax6\"            NA                \"L5/6 NP\"         \"L5 IT\"          \n",
      " [5] \"L6 CT\"           \"L4 IT\"           \"Astrocyte\"       \"L2/3 IT\"        \n",
      " [9] \"Vip\"             \"Sst Chodl\"       \"L6 IT Car3\"      \"L6 IT\"          \n",
      "[13] \"Sncg\"            \"Pvalb\"           \"Oligodendrocyte\" \"Lamp5 Lhx6\"     \n",
      "[17] \"Sst\"             \"VLMC\"            \"Lamp5\"           \"Microglia-PVM\"  \n",
      "[21] \"OPC\"             \"L6b\"             \"Endothelial\"     \"L5 ET\"          \n",
      "[25] \"Chandelier\"     \n"
     ]
    }
   ],
   "source": [
    "mtgseaad <- \"/mnt/nfs/CX000008_DS1/projects/btanasa/brain_refs_MTG_SEA_AD/Reference_MTG_RNAseq_all-nuclei.2022-06-07.sce.rds\"\n",
    "mtg_sea_ad <- readRDS(mtgseaad)\n",
    "\n",
    "# Inspect its structure\n",
    "class(mtg_sea_ad)\n",
    "\n",
    "#.$ cluster_label            : chr [1:166868] \"Pax6_1\" NA \"L5/6 NP_1\" \"L5 IT_7\" ...\n",
    "#.$ subclass_label           : chr [1:166868] \"Pax6\" NA \"L5/6 NP\" \"L5 IT\" ...\n",
    "#.$ class_label              : chr [1:166868] \"Neuronal: GABAergic\" \"NA\" \"Neuronal: Glutamatergic\" \"Neuronal: Glutamatergic\" ...\n",
    "#.$ GA_cluster_label         : chr [1:166868] \"Pax6_1\" NA \"L5/6 NP_1\" \"L5 IT_7\" ...\n",
    "#.$ GA_subclass_label        : chr [1:166868] \"Pax6\" NA \"L5/6 NP\" \"L5 IT\" ...\n",
    "#.$ GA_neighborhood_label    : chr [1:166868] \"lamp5_sncg_vip\" NA \"l5et_l56np_l6ct_l6b\" \"it_types\" ...\n",
    "#.$ CA_cluster_label         : chr [1:166868] \"Pax6_1\" NA \"L5/6 NP_3\" \"L5 IT_1\" ...\n",
    "#.$ CA_subclass_label        : chr [1:166868] \"Pax6\" NA \"L5/6 NP\" \"L5 IT\" ...\n",
    "#.$ CA_neighborhood_label    : chr [1:166868] \"CGE Inh\" NA \"Deep Exc\" \"IT types\" ..\n",
    "\n",
    "# Check available columns in colData\n",
    "colnames(colData(mtg_sea_ad))\n",
    "\n",
    "# Get unique values for each label\n",
    "unique_clusters    <- unique(mtg_sea_ad$cluster_label)\n",
    "unique_subclasses  <- unique(mtg_sea_ad$subclass_label)\n",
    "unique_classes     <- unique(mtg_sea_ad$class_label)\n",
    "\n",
    "# Print counts + first few\n",
    "cat(\"cluster_label:\", length(unique_clusters), \"unique\\n\")\n",
    "print(unique_clusters)\n",
    "\n",
    "cat(\"class_label:\", length(unique_classes), \"unique\\n\")\n",
    "print(unique_classes)\n",
    "\n",
    "cat(\"subclass_label:\", length(unique_subclasses), \"unique\\n\")\n",
    "print(unique_subclasses)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9aae81e4-048d-4da9-b21c-236efce46062",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "'SingleCellExperiment'"
      ],
      "text/latex": [
       "'SingleCellExperiment'"
      ],
      "text/markdown": [
       "'SingleCellExperiment'"
      ],
      "text/plain": [
       "[1] \"SingleCellExperiment\"\n",
       "attr(,\"package\")\n",
       "[1] \"SingleCellExperiment\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Object class: SingleCellExperiment \n",
      "Total features: 36601 \n",
      "Ensembl (ENSG) IDs: 0 \n",
      "Other (likely gene symbols): 36601 \n",
      "\n",
      "Example Ensembl IDs:\n",
      "character(0)\n",
      "\n",
      "Example non-Ensembl IDs:\n",
      "[1] \"MIR1302-2HG\" \"FAM138A\"     \"OR4F5\"       \"AL627309\"    \"AL627309\"   \n",
      "[6] \"AL627309\"   \n"
     ]
    }
   ],
   "source": [
    "# Inspect class\n",
    "class(mtg_sea_ad)\n",
    "\n",
    "# Extract feature IDs, strip version suffixes\n",
    "ids <- sub(\"\\\\.\\\\d+$\", \"\", rownames(mtg_sea_ad))\n",
    "\n",
    "# Flag Ensembl IDs\n",
    "is_ensg <- grepl(\"^ENSG\", ids)\n",
    "\n",
    "# Counts\n",
    "n_total  <- length(ids)\n",
    "n_ensg   <- sum(is_ensg)\n",
    "n_symbol <- n_total - n_ensg\n",
    "\n",
    "cat(\"Object class:\", class(mtg_sea_ad), \"\\n\")\n",
    "cat(\"Total features:\", n_total, \"\\n\")\n",
    "cat(\"Ensembl (ENSG) IDs:\", n_ensg, \"\\n\")\n",
    "cat(\"Other (likely gene symbols):\", n_symbol, \"\\n\")\n",
    "\n",
    "# Peek at examples\n",
    "cat(\"\\nExample Ensembl IDs:\\n\")\n",
    "print(head(ids[is_ensg]))\n",
    "\n",
    "cat(\"\\nExample non-Ensembl IDs:\\n\")\n",
    "print(head(ids[!is_ensg]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "caa4f858-a70d-434c-8f3c-717fae1c3107",
   "metadata": {},
   "outputs": [],
   "source": [
    "# To work with : \n",
    "\n",
    "# cluster_label\n",
    "# subclass_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b000f7ee-4e68-4908-928d-c9e0bffe05fa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
